{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data ingest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### load environment variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from dotenv import load_dotenv\n",
    "\n",
    "load_dotenv()\n",
    "\n",
    "os.environ[\"OPENAI_API_KEY\"]=os.getenv(\"OPENAI_API_KEY\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### load local text file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Document(page_content='We define how chronic cigarette smoke-induced time-dependent epigenetic alterations can sensitize human bronchial epithelial cells for transformation by a single oncogene. The smoke-induced chromatin changes include initial repressive polycomb marking of genes, later manifesting abnormal DNA methylation by 10 months. At this time, cells exhibit epithelial-to-mesenchymal changes, anchorage-independent growth, and upregulated RAS/MAPK signaling with silencing of hypermethylated genes, which normally inhibit these pathways and are associated with smoking-related non-small cell lung cancer. These cells, in the absence of any driver gene mutations, now transform by introducing a single KRAS mutation and form adenosquamous lung carcinomas in mice. Thus, epigenetic abnormalities may prime for changing oncogene senescence to addiction for a single key oncogene involved in lung cancer initiation.', metadata={'source': 'abstract.txt'})"
      ]
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from langchain_community.document_loaders import TextLoader\n",
    "loader=TextLoader(\"abstract.txt\")\n",
    "text_docs=loader.load()\n",
    "text_docs[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### load local text file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "38\n",
      "[Document(page_content=\"Chronic Cigarette Smoke-Induced Epigenomic Changes Precede \\nSensitization of Bronchial Epithelial Cells to Single Step \\nTransformation by KRAS  Mutations\\nMichelle Vaz1, Stephen Y Hwang1, Ioannis Kagiampakis1, Jillian Phallen1, Ashwini Patil2, \\nHeather M O’Hagan3, Lauren Murphy1, Cynthia A Zahnow1, Edward Gabrielson4, Victor E \\nVelculescu1, Hariharan P Easwaran1,5, and Stephen B Baylin1,5,6\\n1Department of Oncology, The Sidney Kimmel Comprehensive Cancer Center, The Johns \\nHopkins University School of Medicine, Baltimore, MD 21287, USA.\\n2Krieger School of Arts and Sciences, Baltimore, MD 21218, USA\\n3Medical Sciences, Indiana University School of Medicine, Bloomington, IN 47405, USA Melvin \\nand Bren Simon Cancer Center, Indianapolis, IN 46202, USA\\n4Department of Pathology, The Johns Hopkins University School of Medicine, Baltimore, MD \\n21287\\nSUMMARY\\nWe define how chronic cigarette smoke-induced time-dependent epigenetic alterations can \\nsensitize human bronchial epithelial cells for transformation by a single oncogene. The smoke-\\ninduced chromatin changes include initial repressive polycomb marking of genes, later \\nmanifesting abnormal DNA methylation by 10 months. At this time, cells exhibit epithelial to \\nmesenchymal changes, anchorage-independent growth and upregulated RAS/MAPK signaling \\nwith silencing of hyper-methylated genes, which normally inhibit these pathways and are \\nassociated with smoking related NSCLC. These cells, in the absence of any driver gene mutations, \\nnow transform by introducing a single KRAS  mutation and form adeno-squamous lung \\ncarcinomas in mice. Thus, epigenetic abnormalities may prime for changing oncogene senescence \\nto addiction for a single key oncogene involved in lung cancer initiation.\\neTOC blurb/In Brief\\n5To whom all correspondence should be addressed: Stephen B. Baylin , Department of Oncology, The Sidney Kimmel Comprehensive \\nCancer Center at Johns Hopkins, sbaylin@jhmi.edu, Hariharan P Easwaran , Department of Oncology, The Sidney Kimmel \\nComprehensive Cancer Center at Johns Hopkins, Baltimore, MD 21287, USA. heaswar2@jhmi.edu.5Lead Contact, Stephen B. Baylin , Department of Oncology, The Sidney Kimmel Comprehensive Cancer Center at Johns Hopkins, \\nBaltimore, MD 21287, USA. sbaylin@jhmi.edu.\\nPublisher's Disclaimer: This is a PDF file of an unedited manuscript that has been accepted for publication. As a service to our \\ncustomers we are providing this early version of the manuscript. The manuscript will undergo copyediting, typesetting, and review of \\nthe resulting proof before it is published in its final citable form. Please note that during the production process errors may be \\ndiscovered which could affect the content, and all legal disclaimers that apply to the journal pertain.\\nAuthor Contributions:\\nM.V ., H.P.E., and S.B.B. designed experiments, performed data analyses, and wrote the manuscript. M.V performed experiments. \\nS.Y .H.. I.K. and A.P. performed bioinformatics analyses under supervision of H.P.E. E.G classified xenograft histopathology. J.P. and \\nV .E.V . performed and analyzed exome sequencing. HOH assisted with data analyses and manuscript review. C.A.Z and L.M assisted \\nwith in vivo experiments.\\nHHS Public Access\\nAuthor manuscript\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nPublished in final edited form as:\\nCancer Cell . 2017 September 11; 32(3): 360–376.e6. doi:10.1016/j.ccell.2017.08.006.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript\", metadata={'source': 'cancercell.pdf', 'page': 0}),\n",
      " Document(page_content='Vaz et al. show that long-term exposure of untransformed human bronchial epithelial cells to \\ncigarette smoke condensate induces epigenetic changes, which are consistent with those \\ncommonly seen in smoking related non-small cell lung cancer, that sensitize the cells to \\ntransformation with a single KRAS mutation.\\nINTRODUCTION\\nIt is well established that chronic exposure to various forms of stress can cause epigenetic as \\nwell as genetic alterations ultimately leading to the development of cancer. Cigarette smoke \\nplays a key role in the development of lung cancer, which remains the leading cause of \\ncancer-related deaths worldwide ( Torre et al., 2015 ). The effect of cigarette smoke and its \\ncomponents in contributing to epigenetic changes in lung cancer is well documented \\n(Belinsky et al., 2002 ; Damiani et al., 2008 ; Liu et al., 2010 ; Tellez et al., 2011 ; Tessema et \\nal., 2014 ). In addition, a number of mutations seen in lung cancer patients are attributed to \\ncigarette smoke exposure ( Cancer Genome Atlas Research, 2012 ; Govindan et al., 2012 ). It \\nis now appreciated that these genetic abnormalities exist with epigenetic changes in all \\nhuman cancers and both presumably contribute to tumorigenesis through induction of \\nabnormal regulation of multiple key signal transduction pathways ( Baylin and Jones, 2011 ; \\nJones and Baylin, 2007 ; Macaluso et al., 2003 ; Shen and Laird, 2013 ; You and Jones, 2012 ). \\nHowever, the exact order for the evolution of these molecular events and their specific \\ncontributions to steps in tumor initiation remains unclear. There are strong suggestions, but \\nlittle direct evidence, that epigenetic changes might lead to altered regulation of key genes \\nand their associated pathways which then play a seminal role in tumor initiation ( Baylin and \\nOhm, 2006 ; Suzuki et al., 2004 ). The direct demonstration of this possibility and the \\nsequential events involved are difficult to study however especially for human cells. For the \\npresent study, we use human bronchial epithelial cells (HBECs), which are initially \\nimmortalized via their having been engineered for overexpression of human telomerase \\nreverse transcriptase (hTERT) and cyclin-dependent kinase 4 (Cdk4) ( Ramirez et al., 2004 ). \\nThe latter engineering causes the CDKN2A  (p16) tumor suppressor gene to be expressed at \\nhigh levels but be unable to perform its normal roles of inhibiting the cell cycle and \\ntriggering cell senescence. However, these cells retain an intact p53 checkpoint, remain Vaz et al. Page 2\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 1}),\n",
      " Document(page_content=\"capable of responding to differentiation signals, are anchorage-dependent and cannot initiate \\ntumor formation in immune-incompetent mice ( Delgado et al., 2011 ; Ramirez et al., 2004 ).\\nMoreover, they require exogenous expression of three or more driver gene mutations for \\ninducing the above abnormal growth and tumorigenic phenotypes ( Sato et al., 2013 ; Sato et \\nal., 2006 ). In this context, our present study directly addresses one hypothesis we have put \\nforth for the early role of abnormal epigenetic events in tumor initiation ( Easwaran et al., \\n2014 ). Namely, these changes could alter signaling to upregulate pathways downstream of \\nkey mutated oncogenes allowing affected cells to subsequently bypass the normal oncogenic \\nsenescence response for the genetic abnormality and rather become addicted to it for \\ntumorigenic effects.\\nRESULTS\\nChronic CSC exposure induces DNA damage-related chromatin binding changes\\nEarlier studies have shown that the transcription repressive proteins DNMT1, EZH2 and \\nSIRT1 bind tightly to DNA at sites of DNA damage following induction of DNA double \\nstrand breaks and/or acute oxidative stress ( O'Hagan et al., 2008 ; O'Hagan et al., 2011 ). We \\ntreated HBECs with a commercially available cigarette smoke condensate (CSC) that is \\nprepared as detailed in STAR methods. CSC concentrations that did not significantly \\ndecrease cell viability were selected based on preliminary dose response curves to define an \\nappropriate concentration for long-term treatment. Treating HBECs with CSC for 10 days, \\nas opposed to DMSO alone, induced chromatin binding of DNMT1, EZH2, and SIRT1. \\n(Figures 1 A–C). While total nuclear protein levels of the maintenance DNA methylation \\nenzyme, DNMT1, increased initially after CSC treatment, the levels decreased by one month \\nand remained decreased for up to 15 months. There was no change in total nuclear levels of \\nEZH2 and SIRT1 (Figures 1A and 1C). Importantly, the tight binding for each of the above \\nproteins increased at 10 days following the acute stress of CSC exposure and then decreased \\nonce cells likely began to adapt to this stress (Figures 1B–1C). The tightness of binding of \\nDNMT1 to chromatin tended to increase again after 10 and 15 months of chronic CSC \\nexposure (Figures 1B and 1C). The expression of TET2, which is involved in DNA \\ndemethylation, is not altered by CSC treatment (Figure S1). Over the time course of \\ntreatment, HBECs underwent distinct and progressive morphological changes, defined in \\nmore detail later below, consisting of initial cell rounding indicative of cells under stress and \\nby 10 months, appearance of more easily visualized and flattened cells (Figure 1D).\\nChronic exposure to CSC causes progressive, genome-wide alterations in DNA \\nmethylation\\nHBEC cells chronically exposed to CSC evolve dramatic and progressive alterations in DNA \\nmethylation across the genome. These include, simultaneously, both global losses and more \\nfocal methylation gains, with the latter especially involving normally unmethylated promoter \\nCpG islands. These time dependent dynamics differ starkly from, and are far more extensive \\nthan those seen in untreated control cells grown on plastic for the same time periods. Thus, \\nin principal component analysis (PCA) plots for two independent biological replicates of \\neach, control cells have relatively tightly clustered methylation changes throughout the 15 Vaz et al. Page 3\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript\", metadata={'source': 'cancercell.pdf', 'page': 2}),\n",
      " Document(page_content='months of growth while the CSC treated cells progressively vary significantly from these \\nand also show, especially after 6 months, greater variation between their replicates (Figure \\n2A and 2B). For the latter cells, these changing methylation patterns suggest many probable \\nstochastic events occurring heterogeneously (Figures 2A–C). Thus, a time dependent change \\nin methylation is seen following CSC treatment and stark changes are observed in both \\nreplicates following 10 months of exposure (Figure 2C). Methylation changes do occur, but \\nare far fewer and occur in a more coordinated pattern in control cells with little variation \\nbetween replicates as compared to CSC treated cells (Figure 2B). The patterns in the CSC \\ntreated cells also indicate that while the replicate patterns diverge continually from each \\nother, and from control cells, the 10 and 15 month time point patterns for each replicate \\nremain clustered together (orange circles versus green circle - Figure 2A).\\nAt CpG islands within ± 1500 bp from transcription start sites (TSS), CSC treated cells have \\nvirtually all gains in DNA methylation with time as observed by the increase in the number \\nof low basal methylated probes in section ‘a’ over the time course of the treatment (Figure \\n2D - red probes and  S2A - red bars ). In contrast, gene body regions predominantly \\nunderwent hypomethylation with time in CSC treated cells as seen by the increase in \\nnumber of probes in section ‘d’ (Figure 2D - blue probes and  S2A - blue bars ). Putative \\nenhancer regions showed both gains and losses in methylation with a trend towards the latter \\n(Figures 2D and S2A). Methylation changes in the “shore regions” which lie just outside the \\nTSS-associated CpG islands resembled changes in both TSS and gene body regions showing \\nboth gains and losses of methylation, whereas there were slight losses in methylation in the \\nshelf regions lying distal to the shores (Figure S2B).\\nDistinct categories of genes become abnormally methylated at promoters with chronic \\nexposure to CSC\\nAnalysis of top variable TSS associated CpG island probes with low basal methylation \\nlevels in the control cells passaged at 1 month, revealed that they grouped into three \\ncategories (Figure 3A). The first group (group I) contained probes that were slightly \\nmethylated in controls over time and further methylated by CSC. These changes strikingly \\nresemble the promoter DNA methylation changes affecting genes in human cancers ( Issa, \\n2014 ; Maegawa et al., 2014 ). The second group (group II) exhibited de novo methylation \\nonly in CSC treated cells after 6 months. The third group (group III) gained methylation \\nonly in the control cells and remained unmethylated with CSC treatment. To establish the \\nrole of the CSC mediated methylation changes, we obtained a list of CSC-specific \\nmethylated genes (Table S1) from both replicates at 6, 10 and 15 months relative to control \\ncells at the same times points using the criteria detailed in the STAR methods section. \\nFurther categorization of these genes by Gene Ontology analysis shows them to be involved \\nin stem cell fates concerned with development, cell fate decisions and embryonic \\nmorphogenesis (Figure 3B). This is similar to what has been found for the majority of \\npromoter hypermethylated genes in cancer ( Easwaran et al., 2012 ; Helman et al., 2012 ). The \\nindividual genes fall into the functional categories of homeobox genes, transcription factors \\nas well as genes involved in glycosylation. These gene categories remained constant at all \\ntime points, although the number of genes in each category increased with time (Table S2). \\nImportantly, there is a very significant overlap between these CSC-specific hypermethylated Vaz et al. Page 4\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 3}),\n",
      " Document(page_content='genes and genes methylated in both lung adeno- (LUAD) and squamous cell (LUSC) \\ncarcinomas (Figure 3C and Table S2). Additionally these methylation changes are also \\nobserved in other oncologic states such as cancers of the breast, colon and pancreas (Figure \\nS3A). Genes hypermethylated in the above cancers were obtained by selecting probes \\nmethylated in at least 25% of tumors and identifying corresponding genes. In a further link \\nfor all the above changes to diseases associated with chronic cigarette smoking, there is a \\nmoderate, but statistically significant, increase in methylation of our CSC-methylated genes \\nin lung tissues obtained from people with chronic obstructive pulmonary disease (COPD) \\nand idiopathic lung fibrosis (Figures S3B and S3C). Finally, in addition to promoter-\\nassociated hypermethylation, CSC exposure also induced much less frequent \\nhypomethylation events in these regions (Figures 2D, S2A and S3D). Categorization of these \\ngenes shows them to be involved in cell signaling as well as negative regulation of cell \\ngrowth (Figure S3E).\\nDynamic changes in chromatin marks and binding of chromatin-associated proteins \\nprecede DNA methylation\\nA building hypothesis for the evolution of promoter DNA hypermethylation in cancer is that \\nthe genes involved are primarily those associated with transcription regulation by polycomb \\ngroup proteins (PcG) and not by promoter DNA methylation in normal tissues. These are \\ntypically developmental genes with unmethylated promoter CpG islands that are surrounded \\nby bivalent chromatin, characterized by simultaneous presence of the repressive mark, \\nH3K27me3 and the active mark, H3K4me3 around the TSS ( Azuara et al., 2006 ; Bernstein \\net al., 2006 ) and associated with low basal, but poised, expression in embryonic (ESC) and \\nadult stem cells. It is hypothesized that during tumor initiation and progression, a \\nquantitative shift occurs away from this chromatin state towards a more repressive one \\nmarked by DNA methylation ( Easwaran et al., 2012 ; Ohm and Baylin, 2007 ; Schlesinger et \\nal., 2007 ; Widschwendter et al., 2007 ). Yet, there are no data from dynamic models to \\nconfirm this. We investigated whether such a series of events occurs during chronic exposure \\nof HBEC cells to CSC.\\nIn initial ChIP studies in CSC exposed versus control cells, we followed enrichment of \\nH3K4me3 and H3K27me3, as well as binding of DNMT1 and EZH2, at proximal promoter \\nCpG island regions of candidate tumor suppressor genes ( BMP3 , SFRP2 , GATA4 ) known to \\nbe frequently methylated in lung and other cancers. By 10 days of CSC exposure, these \\ngenes accumulate bivalent chromatin with H3K4me3 and H3K27me3 marks at their \\npromoters (Figure 4A). In stark contrast, genes that are not hypermethylated in cancer, \\nincluding ACTB,  a house keeping gene with a CpG island containing promoter, and \\nNANOG,  an embryonic related gene with a CpG poor promoter, do not develop this change. \\nWhile the promoter protein binding and chromatin marks studied do not change for the latter \\ntwo control genes over the entire duration of CSC exposure, BMP3 , SFRP2  and GATA4 \\nexhibit initial increases in binding of the enzyme catalyzing H3K27me3, EZH2 (Figure 4A). \\nOver the time course of CSC exposure and associated with evolution of promoter DNA \\nhypermethylation, there is a quantitative switch from the presence of EZH2 to binding of the \\nmaintenance protein for DNA methylation, DNMT1, and a concomitant decrease in levels of \\nH3K27me3, and the active mark, H3K4me3, at the 15 month time point (Figure 4A). While Vaz et al. Page 5\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 4}),\n",
      " Document(page_content='expression levels of the DNA hypermethylated genes remain unchanged at initial time points \\nfollowing CSC treatment, their expression decreases following 10 months of CSC treatment \\ntime (Figure 4B).\\nThe above studies are well reflected in genome wide ChIP-seq studies (Figures 4C and \\nS4A–G), which assessed enrichment of the above mentioned histone marks/epigenetic \\nproteins at promoter CpG islands of genes (picked from Table S1), which progressively \\nbecome DNA hypermethylated during chronic CSC exposure. This is best appreciated when \\nthese genes are compared to a control set of random CpG island containing genes not \\nmethylated by CSC. First, these latter genes have a much higher average basal expression \\n(Figure S4H) and importantly, a distinct, clear zone at the TSS indicating lack of DNMT1 \\nbinding in this region (Figure 4C -bottom right ). In contrast, the genes that get \\nhypermethylated by CSC have a much lower average basal expression (Figure S4H), and \\nunder both control and CSC exposure conditions, DNMT1 binding encroaches upon the TSS \\nregion (Figure 4C). The levels of DNMT1 around the TSS in these genes initially decreases \\nduring acute CSC exposure (10 days) compared to the control cells, but later increases over \\ntime in the CSC exposed cells as seen in the plots showing average read counts of DNMT1 \\nenrichment in CSC-methylated genes (Figure 4C -bottom right ). This initial decrease in \\nDNMT1 levels may reflect an opening of the chromatin as the acute CSC exposure induces \\nan increase in promoter bivalency as seen by the increase in the H3K4me3 mark in CSC \\ncells at 10 days (Figure 4C -top left ), described further below.\\nSimilar to the chromatin changes observed in ChIP studies for candidate genes, ChIP-seq \\nanalysis of CSC methylated genes at 10 days also showed increases in EZH2 (Figure 4C -top \\nright ) and H3K4me3 (Figure 4C -top left ) around the TSS as compared to the control cells. \\nThe bivalent nature of the CSC-methylated genes is well appreciated when these genes are \\ncompared to the control set of random unmethylated genes which have far higher levels of \\nH3K4me3 and virtually absent H3K27me3 (Figure 4C- top and bottom left ). By 10 months, \\nthe CSC-methylated genes, despite maintaining distinct H3K27me3 marks in both control \\nand CSC exposed cells, have far lower levels of EZH2 than in control cells which show a \\nsharp increase in EZH2 levels with time (Figures 4C- top right and  S4A–G). All of the \\nabove results indicate that chronic CSC exposure induces a dynamic, time-dependent switch \\nfrom promoter PcG control to promoter DNA methylation for a group of low expression \\ngenes with initial promoter region bivalent chromatin. Interestingly, the increased EZH2 in \\nthe promoter region for these genes seen in control cells may indicate that their low \\nexpression controlled by PcG may facilitate long-term growth as immortalized cells. The \\nramifications of these changes are evident, as outlined below in gene expression and \\nphenotypic differences between control and CSC exposed cells.\\nProgressive changes in oncogenic pathway signaling are induced by CSC exposure in the \\nabsence of driver gene mutations\\nThe epigenetic changes chronicled above accompany CSC induced alterations of signaling \\npathways. Initially, at 10 days and 1 month of CSC exposure upregulation for cell cycle and \\nDNA replication events dominated followed by pathways for repair of DNA damage and \\nextracellular environment interactions (Figure 5A and B). These changes may reflect Vaz et al. Page 6\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 5}),\n",
      " Document(page_content='adaptation of exposed cells to survive the initial stress induced by CSC treatment. \\nSubsequently, with continued CSC exposure and consistent with the chromatin changes \\noutlined earlier, there are switches in dominant pathway signaling at 3 and 6 months after \\nCSC exposure including downregulation of gene targets of the PcG proteins BMI1 and \\nEZH2 (Figure 5A and B). Thereafter, when abnormal DNA methylation is the dominant \\nepigenetic change by 10 to 15 months of CSC treatment, top pathway signaling switches to \\nactivation of key potential transformation promoting pathways. Thus, there is strong \\ninduction of KRAS and its downstream MAPK signaling, followed by WNT and EGFR \\nsignaling (Figure 5A, 5B and S5). Importantly, exome sequencing at these time points \\nreveals no mutations for key oncogenic driver genes as discussed later (Table S3).\\nKey gene hypermethylation changes are linked to the above upregulation of the KRAS and \\nWNT pathways (Tables 1 and S2), and importantly, these genes are also frequently \\nhypermethylated, as mentioned earlier in LUAD and LUSC (Figure 3C and Table S2). First, \\nis BMP3 , a member of the TGF- β superfamily, which normally represses RAS-MAPK \\nsignaling ( Kraunz et al., 2005 ). This gene is very frequently promoter hypemethylated in \\nNSCLC, and these events highly correlate with the presence of KRAS  mutations ( Dai et al., \\n2001 ; Kraunz et al., 2005 ). Second, CSC related hypermethylation of a series of genes, \\nincluding MSX1, WIF1  and associated anti-WNT genes have important functional correlates \\n(Table 1). In neural cells, MSX1 counters the upregulation of the WNT pathway ( Revet et \\nal., 2008 ; Revet et al., 2010 ), the other leading oncogenic pathway upregulated in our studies \\n(Figure 5 and S5). MSX1 induces at least four different WNT pathway inhibitor genes: \\nDickkopf 1-3 ( DKK1-3 ) and secreted frizzled-related protein 1 ( SFRP1 ) and SFRP2  (Revet \\net al., 2010 ). These are all also hypermethylated in the late CSC exposed cells and also \\nfrequently in primary lung cancers.\\nCSC exposure induces time dependent development of EMT, anchorage independent \\ngrowth and susceptibility to KRAS -induced transformation\\nTracking with the progression of epigenetic changes induced by CSC, the HBEC cells \\nmanifest key changes suggesting a drift towards a pre-tumorigenic state. First, by 10 to 15 \\nmonths, the cells become anchorage independent and form clones in soft agar (Figure 6A \\nand 6B). At this time point, when these CSC exposed cells are grown in 5% FBS containing \\nmedium, they become more elongated, suggestive of an EMT like phenotype (Figure 6C), \\nand this is accompanied by a stark increase in vimentin, a key protein upregulated in EMT, \\nand a decrease in E-cadherin (Figure 6D and 6E). However, despite these changes, neither \\nthe CSC-exposed cells nor the isolated soft agar clones could form tumors in vivo.\\nWe hypothesized that the above progressive development of the epigenetic, signaling \\npathway and morphogenic changes induced by CSC exposure might sensitize the cells to \\ntransformation by a single oncogenic event. This would be highly significant since earlier \\nstudies have shown that complete transformation of HBEC cells requires introduction of at \\nleast three genetic events: shRNA-mediated downregulation of TP53 paired with \\nintroduction of both KRAS  mutations and CMYC overexpression ( Sato et al., 2013 ). We \\nintroduced individually either KRASV12 (Figure 6F) or downregulate the tumor suppressor \\nTP53 (Figure 6G) in 6 and 15 month CSC treated cells. Only KRASV12 alters the 15 month Vaz et al. Page 7\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 6}),\n",
      " Document(page_content='CSC-treated cells, while neither impact the 6 month CSC-exposed cells or their respective \\ncontrols. There is a dramatic increase in anchorage-independent growth in vitro  with more \\nnumerous, enlarged and irregular colonies in CSC exposed cells expressing mutant KRAS \\n(Figures 6H and 6I). These cells continue to have increased KRAS and WNT signaling now \\njoined by significant enrichment of MTOR, Hedgehog and VEGF signaling pathways \\n(Figure 6J). Most important, with exogenous expression of KRASV12, but not TP53 \\ndownregulation, the CSC treated cells are now transformed. Neither of the genetic \\nmanipulations transformed control cells. KRASV12 - expressing 15 month CSC exposed \\ncells from both replicates induce tumors at different rates in mice which appear 3 weeks post \\ninjection and grow to an average size ranging from 500–900 mm3 (Figure 6K and 6L). \\nExpression of mutant KRASV12 was detected in the mouse xenografts but not in parental \\nCSC cells transfected with empty vector (Figure 7A). Tumor histology suggests the tumors \\nhave an adeno-squamous phenotype with sheets of poorly differentiated cells and focal areas \\nof both squamous and glandular differentiation (Figure 7B). The squamous differentiation \\nwas confirmed by positive staining for p63 and cytokeratin (CK) 5/6. Although tumor cells \\nstained negative for TTF1 (NKX2-1), a commonly used marker for pulmonary \\nadenocarcinoma, these were positive for CK7, which is more commonly expressed in human \\npulmonary adenocarcinoma than squamous cell carcinoma ( Camilo et al., 2006 ). Areas of \\ntumor with glandular differentiation stained for mucin, further indicating their mixed adeno- \\nand squamous differentiation.\\nMethylation analysis of control and CSC cells following KRASV12 expression, as well as in \\nxenografts, showed in both replicates there is maintenance of and slight increases in the \\nabnormal promoter methylation of the genes classified as CSC-methylated at 15 months \\n(Table S1, Figures S6A and S6B). Moreover, analysis of methylation at promoters of a key \\nset of CSC-methylated genes (Table S2), revealed an even greater increase in promoter \\nmethylation of these genes, accompanied by a decrease in their expression in the xenografts \\n(Figure 7C, 7D, S6C and S6D). Gene expression analysis in xenografts and in CSC treated \\ncells expressing KRASV12 showed enrichment of KRAS, WNT and Hedgehog pathways \\n(Figure 7E). Upregulation of LEF1 signaling suggests an important role of WNT pathway in \\ndriving tumor formation as WNT signaling through LEF1 mediates metastasis of lung \\nadenocarcinoma ( Nguyen et al., 2009 ). These results all underscore that stress induced \\nepigenetic alterations induce signaling events that prime cells for transformation and \\noncogenic addiction, rather than induction of senescence by a single oncogene. In this \\nregard, whole exome sequencing of > 20,000 genes in the parental HBEC cells, control and \\nCSC treated cells at 15 months and tumor xenografts, with an average sequence coverage \\nwithin the target regions of nearly 150-fold for each sample reveals the absence of well-\\nknown mutated driver genes for NSCLC and other cancers except for our introduced \\nKRASV12 mutation (Table S3). While there was a median of 72 mutations in the 15 month \\nCSC treated cells, 114 in the xenografts, and 62 in the 15 month DMSO treated control cells \\nthat were not present in the parental HBEC cells, these all likely reflect passenger mutations \\nthat arose in the cells prior to xenograft formation. To exclude the presence of rare clones \\nwith driver mutations, we also performed targeted deep sequencing of three xenografts with \\n> 600-fold coverage at each base using a 123 gene panel (Table S3 - targeted genes ) and did \\nnot find any additional mutations not previously detected by exome sequencing (Table S3). Vaz et al. Page 8\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 7}),\n",
      " Document(page_content='We observed in the exome and targeted analyses alterations in JAK1  and CREBBP  genes, \\nbut mutations at the affected amino acid residues are rare in all cancers and not likely to \\nhave biological consequences.\\nCSC induced methylation changes associate with smoking related non-small cell lung \\ncancers with a bias towards those with KRAS mutations\\nThe above bias to KRAS-mediated transformation raises interesting questions with respect \\nto smoking and the mutation distributions among sub-groups of primary NSCLC. First, \\nanalyses of TCGA data from lung adenocarcinomas ( Cancer Genome Atlas Research, 2014 ) \\nsuggest that the epigenetic changes in our model may well sensitize bronchial cells in \\nhumans to smoking related tumorigenesis, which accounts for nearly 80–90% of NSCLC. \\nLower methylation of our CSC-specific methylated genes are present in lung \\nadenocarcinomas with EGFR  mutations compared to tumors with KRAS  or TP53  mutations \\n(Figure 7F). It has been shown that EGFR  mutant tumors very frequently occur in non-\\nsmokers unlike KRAS  and TP53  mutated tumors which frequently occur in smokers ( Boch \\net al., 2013 ). Second, consistent with the specificity for KRAS  mutations outlined in the \\nprevious section, these methylation changes are higher in matched normal lung samples \\nfrom KRAS  mutated tumors than matched normal samples from TP53  mutated tumors \\n(Figure 7F). Third, although the methylation of our genes are comparable between TP53  and \\nKRAS  mutant tumors, we see interesting differences between duration of smoking and the \\nstatus of smokers (current versus reformed). Our CSC-specific methylated genes are more \\nhighly methylated in primary lung tumors from current smokers as compared to tumors from \\nnon-smokers or reformed smokers (Figure 7G). However, particularly intriguing is that, in \\ntumors from patients who were reformed smokers for 15 years or more, there are starkly \\nlower methylation levels of our CSC-specific hypermethylated genes in TP53  mutant tumors \\nversus those with KRAS  mutations. This may indicate that the presence of a KRAS  mutation \\nleads to more maintenance of methylation for our CSC methylated genes than for tumors \\nharboring TP53  mutations without KRAS  mutations. We hypothesize that although the \\nchanges seen in our model may correlate with smoking induced sensitization to multiple \\nmutations in all NSCLC, TP53  mutations may need to occur with additional genetic \\nalterations. KRAS  mutations on the other hand may more readily accomplish one-step \\ntransformation in the setting of the epigenetic and signaling changes we observe with CSC \\nexposure.\\nDISCUSSION\\nWe have derived data suggesting how a progressive increase in epigenetic abnormalities \\nduring chronic cigarette smoke exposure may drive the initiation of lung cancer. This \\nexposure pushes cells to more stem cell like chromatin states associated with signaling \\nevents sensitizing the cells to oncogene induced transformation. In this regard, epigenetic \\nabnormalities now recognized in scenarios of chronic cellular stress, inflammation, and \\naging are suggested, but not proven, to drive early stages for cancer development ( Issa, \\n2014 ; Maegawa et al., 2014 ; Niwa and Ushijima, 2010 ). Defining how these events interact \\nwith key genetic alterations, like KRAS , is an important goal for cancer research. Our \\ncurrent data are consistent with our previous hypothesis that the types of epigenetic Vaz et al. Page 9\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 8}),\n",
      " Document(page_content=\"abnormalities observed in the present study could provide an “epigenetic substrate” allowing \\ncells to become addicted to, rather than senescing or dying, in responding to signaling from \\nmutated driver genes, like KRAS  (Baylin, 2012 ; Dimauro and David, 2010 ).\\nOne key implication of our data is that smoke-induced stress results in a more stem-like state \\nin the HBECs facilitating transformation by a key oncogenic event. Other studies report that \\nchronic stress could dedifferentiate mature cells to a more stem-like state ( Lee et al., 2015 ; \\nStange et al., 2013 ; Tata et al., 2013 ). It has been observed that p53 or BRAF can exhibit \\ntheir tumor driving influence only in an epigenetically controlled stem-cell state ( Kaufman \\net al., 2016 ). Thus, our early, CSC induced epigenetic changes and other environmental \\nfactors could induce cell signaling pathways that cooperate with gene mutations in driving \\ntumorigenesis.\\nThe nature of the observed epigenetic abnormalities are very consistent with key elements of \\nthe above hypotheses. The initial chromatin changes, specifically tightening of DNMT1 and \\nEZH2 to chromatin following a 10 day exposure to a low dose of CSC, are of the type we \\nhave previously associated with double strand DNA breaks and/or ROS exposure ( O'Hagan \\net al., 2008 ; O'Hagan et al., 2011 ). The decrease in nuclear DNMT1 corresponds with \\npreviously published data ( Liu et al., 2010 ), but our results emphasize that it is tight binding \\nto chromatin which is central to the cell stress scenario. This places DNMT1 strategically for \\nmediating maintenance of abnormal DNA methylation induced by chronic CSC exposure. \\nThe increased DNMT1 binding to chromatin at 10 days post exposure and no change over \\nthe next 3–6 months, followed by an increase after 10 months is consistent with events for \\nhow the abnormal gene promoter DNA methylation evolves and suggests possible clonal \\nselection of cells with this abnormal methylation during CSC exposure.\\nSeveral key features of our data link lung cancer and the CSC-induced abnormally DNA \\nmethylated genes. First, the genes overlap significantly with those commonly \\nhypermethylated in NSCLC. Second, the increased methylation levels of these genes in \\nadenocarcinomas from smokers versus reformed or non-smokers suggests their key role in \\nthe development of smoking-associated lung cancer. Third, as discussed earlier, our ChIP \\ndata at selected candidate genes and the genome-wide ChIP-seq studies are in line with \\nhypothesized events for how abnormal promoter DNA methylation may reflect a molecular \\nprogression from polycomb to DNA methylation control. The involved genes are heavily \\nbiased to developmental and stem cell control, which appear most vulnerable to evolve \\nabnormal promoter DNA methylation during cancer initiation and progression ( Helman et \\nal., 2012 ).\\nAn aspect of our data is linking the above vulnerability of involved genes to evolve such \\nmethylation to impingement of basal distribution of DNMT1 near the TSS as compared to \\nhighly transcribed genes with unmethylated promoter CpG islands. This suggests a strategic \\npositioning for DNMT1 to mediate maintenance of promoter DNA methylation evolving \\nwith tumor initiation and/or progression. Importantly, DNMT’s have scaffolding functions, \\nwhich can exert transcriptional repressor functions ( Clements et al., 2012 ; Fuks et al., 2000 ; \\nHa et al., 2011 ; Rountree et al., 2000 ). In addition, polycomb system components can recruit \\nDNMT’s to cancer-specific methylated promoters with simultaneous repression of their Vaz et al. Page 10\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript\", metadata={'source': 'cancercell.pdf', 'page': 9}),\n",
      " Document(page_content='expression even prior to the appearance of abnormal DNA methylation ( Mohammad et al., \\n2009 ).\\nPerhaps most importantly, we now suggest how the above epigenetic abnormalities help \\nCSC exposed cells acquire transformed phenotypes. The augmentation of the KRAS/MAPK \\nand WNT signaling pathways concomitant with peak abnormal gene promoter methylation \\nat 10 months implies the involved genes are key for driving these pathways. Indeed, \\nfunctional studies have shown these genes are important for normally preventing abnormal \\nRAS-MAPK ( Kraunz et al., 2005 ) and WNT pathway signaling ( Revet et al., 2008 ; Revet et \\nal., 2010 ) (Mazieres et al., 2004 ; Nojima et al., 2007 ). Overexpression of these genes in \\ncancer cells in which they are epigenetically silenced causes apoptosis and downregulation \\nof key target genes such as c-MYC ( Cowling et al., 2007 ; Suzuki et al., 2004 ).\\nOur model data may relate to the actual course of lung cancer initiation and progression. \\nOne caveat for our model system is the cells utilized are already immortalized and the \\nengineered loss of function of p16 would induce senescence bypass. However, we do not \\nbelieve this negates our data for the temporal order of the other events we observe vis a vis \\ntheir importance for the genesis of lung cancer. We stressed earlier that others have shown \\nthat HBEC cells need a combination of at least three oncogenic events to form tumors in \\nmice ( Sato et al., 2013 ; Sato et al., 2006 ). Indeed, insertion of KRAS  mutations at 6 months \\nof CSC exposure was not sufficient to transform the cells, consistent with the CSC-induced \\nmethylation changes playing a key role in enhancing susceptibility to one-step KRAS-\\ninduced transformation. We also hypothesize that the fifteen month time span in our studies \\nmimics the typical long course of smoking history for patients developing NSCLC. We have \\npreviously hypothesized that epigenetic abnormalities, such as those induced by our CSC \\nexposure provide a molecular environment facilitating cellular addiction to driver genetic \\nevents and our current findings support this for NSCLC ( Baylin, 2012 ). Our outlined \\nrelationships of CSC-hypermethylated genes in primary lung tumors with mutant KRAS  and \\nTP53  versus EGFR  mutations and with smoking history further suggest the importance of \\nthis epigenetic change in evolution of smoking associated NSCLC. Our modeled events may \\nsensitize to mutations in addition to KRAS  for NSCLC initiation and progression. However, \\nthey may have special relevance for helping KRAS  mutations transform in a single step. \\nActivating RAS/MAPK and WNT pathways are important for cigarette smoke induced lung \\ncancer ( Lemjabbar-Alaoui et al., 2006 ) (Stewart, 2014 ). It is also suggested that KRAS \\nmutations alone are not a strong predictor for developing lung cancer, but associated \\nincreases in WNT signaling accelerate lung tumorigenesis ( Pacheco-Pinedo et al., 2011 ; \\nPacheco-Pinedo and Morrisey, 2011 ). Finally, our findings stress that therapies to revert the \\ntypes of epigenetic abnormalities we have now chronicled could be one key modality for \\npreventing lung cancer in high risk individuals.\\nSTAR METHODS\\nCONTACT FOR REAGENT AND RESOURCE SHARING\\nFurther information and requests for resources and reagents should be directed to and will be \\nfulfilled by the Lead Contact, Stephen Baylin (sbaylin@jhmi.edu).Vaz et al. Page 11\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 10}),\n",
      " Document(page_content=\"EXPERIMENTAL MODEL AND SUBJECT DETAILS\\nCell culture and Treatments— Human bronchial epithelial cells (HBEC-3KT) were \\nobtained from Drs. Shay and Minna, Southwestern Medical Center, Dallas, TX ( Ramirez et \\nal., 2004 ). Cells were maintained at 37°C in the presence of 5% CO2 in keratinocyte serum-\\nfree medium (KSFM; Life Technologies Inc.) media containing 50 mg/mL of bovine \\npituitary extract (BPE; Life Technologies Inc.) and 5 ng/mL of EGF (Life Technologies Inc.) \\n(Ramirez et al., 2004 ).\\nCells were passaged every 2–3 days and fresh CSC and DMSO were added to the CSC and \\ncontrol treated cells respectively at each passage. Cells were treated with 5 µg/ml and 10 \\nµg/ml CSC (Murthy pharmaceuticals) or corresponding concentrations of DMSO (Controls). \\nTwo independent sets of treatments (replicate 1 and replicate 2) were done to obtain the two \\ntechnical repeats that were used for all studies. To determine if CSC treated cells exhibited \\nEMT like characteristics, CSC treated cells at 10 and 15 months along with their respective \\ncontrols were grown in medium with 5% FBS for 96 hr as described earlier ( Sato et al., \\n2013 ).\\nIn Vivo Mouse Studies— We used male NSG mice (5–6 weeks old) (Jackson \\nlaboratories) for these studies. Mice were housed at the Johns Hopkins Animal care facility. \\nAll animal experiments were approved by the Johns Hopkins Animal Care and Use \\nCommittee. All animal care and protocols followed were in accordance with guidelines of \\nthe institutional Animal Care and Use Committee (IACUC). Mice were injected \\nsubcutaneously in the flank with 2 × 106 viable CSC and control cells obtained from the 15 \\nmonth treated cells from both replicates expressing KRASV12 or shTP-53 in 0.2 mL of PBS \\nwith matrigel. Mice were monitored every 3 days for tumor formation for up to 5 months.\\nMETHOD DETAILS\\nCigarette smoke condensate— CSC was purchased from Murthy pharmaceuticals \\n(Lexington, KY , USA). According to the manufacturer’s brochure, the condensate is \\nobtained by smoking University of Kentucky's Standard Research cigarettes (3R4F; \\nUniversity of Kentucky, KY , USA) on an FTC Smoke Machine. The total particulate matter \\nwas collected on Cambridge glass fibre filters and the amount obtained determined by \\nweight increase of the filter. CSC was prepared by dissolving the collected smoke \\nparticulates in dimethyl sulfoxide (DMSO) to yield a 4% solution (w/v).\\nNuclear Protein Extraction and Tight Chromatin Fractionation— CSC or control \\n(DMSO treated) cells were collected at the indicated time points. To prepare nuclear \\nextracts, cell pellets were were lysed in CEBN buffer (10 mM HEPES [pH 7.8], 10 mM \\nKCl, 1.5 mM MgCl2, 0.34 M sucrose, 10% glycerol, 0.2 % NP-40, 1X protease inhibitor \\ncocktail (Invitrogen), 1X phosphatase Inhibitor cocktail, N-ethylmaleimide (Sigma) and \\nPefabloc SC AEBSF (Roche Applied Science)) and pellets were washed with CEB buffer \\n(CEBN buffer without NP-40). For preparation of total nuclear protein extracts, nuclear \\npellets were lysed with 4% SDS and passed through QIAshredder (Qiagen). For preparation \\nof Tight Chromatin fractions, nuclear pellets obtained after washing with CEB buffer, were \\nfurther washed with soluble nuclear buffer (3 mM EDTA, 0.2 mM EGTA, inhibitors as Vaz et al. Page 12\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript\", metadata={'source': 'cancercell.pdf', 'page': 11}),\n",
      " Document(page_content='described above), salt buffer with 0.45M NaCl concentration (50 mM Tris pH 8.0, 0.05% \\nNP40, 0.45M NaCl). The pellet was then lysed in 4% SDS buffer using a Qiashredder \\ncolumn (Qiagen) and referred to as tight chromatin. Western blots of the protein extracts \\nwere probed with DNMT1 (Sigma), EZH2 XP (Cell Signaling), SIRT1 (Delta Biolabs), \\nLamin B (Santacruz Biotechnology). Band densitometry for western blots were quantitated \\nusing ImageJ software.\\nGenome-wide DNA-methylation analysis— Genome-wide DNA methylation was \\nstudied in the control and CSC treated cells at 1, 6, 10 and 15 months as well as in the 15 \\nmonth control and CSC treated cells expressing empty vector (EV) or KRASV12 and the \\ntumor xenografts . These samples were collected from both replicates. DNA was extracted \\nfrom cells and xenograft tissues using the Qiagen DNeasy blood and tissue kit (Qiagen, \\nCA).The Infinium HumanMethylation450k BeadChip array (Illumina) ( Bibikova and Fan, \\n2009 ) was used to analyze bisulfite-treated DNA (EZ DNA-Hypermethylation Kit, Zymo \\nResearch) as previously described ( Easwaran et al., 2012 ). The R Bioconductor package \\nminfi was used to process the raw IDAT files. Probes with a detection p value > 0.01 and \\nprobes associated with sex chromosomes were removed after the data was SWAN-\\nnormalized. Β values were computed as the signal of the methylation-specific probe over the \\nsum of the signals of the methylation- and unmethylated-specific probes. The data is then \\nsubsetted into further groups: cpg islands, body, shelf and shore probes using the annotation \\ninformation available from Bioconductor IlluminaHumanMethylation450k.db library. To \\nfurther get a comprehensive list of all probes that are around the TSS, the CpG coordinates \\nfor the Infinium 450K array were obtained from IlluminaHumanMethylation450k.db and \\nmapped to ± 1500 bp around all transcripts in hg19 (UCSC) using R package BiomaRt \\n(Durinck et al., 2009 ). Probes around the TSS were split by whether or not they are within a \\nCpG island (obtained from hg19 genome from UCSC ( https://genome.ucsc.edu/ )) to obtain \\nthe CpG-island and non-CPG island containing TSS/promoter probes. Probes in putative \\nenhancer regions were obtained by selecting for probes from Infinium \\nHumanMethylation450k BeadChip array that matched with enhancer regions identified in \\nlung cancer in an earlier published study ( Aran et al., 2013 ). This list was further filtered to \\nexclude probes that were within ± 2 kb of a TSS. Scatter plots to examine methylation \\nchanges occurring between control and CSC treated cells at each time point were generated \\nby plotting CpG-island probes mapped to ± 1500bp of TSS.\\nTo obtain a list of CSC methylated genes, we first chose probes that had a starting β value of \\n≤ 0.2 in the untreated cells at 1 month. We then filtered for those probes whose β values did \\nnot increase by more than 0.2 between the control at 1M and the respective control end point \\nand showed an increase of at least 0.3 in β values in CSC treated cells as compared to β \\nvalues in the control cells at the respective end point (6 months, 10 months and 15 months).\\nHypermethylated gene lists for each time point of replicates 1 and 2 were generated using \\nthe following equations:\\n6 months: Con(1 M) ≤ 0.2 AND (Δ β Con (6 M-1 M) ≤ 0.2 and Δ β CSC 6 M - Con 6 \\nM ≥ 0.25),Vaz et al. Page 13\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 12}),\n",
      " Document(page_content=\"10 months: Con(1 M) ≤ 0.2 AND (Δ β Con (10 M-1 M) ≤ 0.2 and Δ β CSC 10 M - \\nCon10 M ≥ 0.25),\\n15 months: Con(1 M) ≤ 0.2 AND (Δ β Con (15 M - 1 M) ≤ 0.2 and Δ β CSC 15 M-\\nCon 15 M ≥ 0.25).\\nHypomethylated genes lists for each time point of replicates 1 and 2 were generated using \\nthe following equations:\\n6 months: Con(1 M) ≥ 0.5 AND (Δ β Con (6 M -1 M) > - 0.2 AND Δ β CSC 6 M-Con \\n6 M ≤ −0.2\\n10 months: Con(1 M) ≥ 0.5 AND (Δ β Con (10 M -1 M) > - 0.2 AND Δ β CSC 10 M-\\nCon 10 M ≤ - 0.2\\n15 months: Con(1 M) ≥ 0.5 AND (Δ β Con (15 M - 1 M) > - 0.2 AND Δ β CSC 15 M \\n- Con 10 M ≤ - 0.2\\nHeatmaps with methylation data— Heat maps are based on hierarchical clustering of β \\nvalues using Euclidean distance and Ward's algorithm (R package gplots). Heatmaps for \\nhypermethylated probes were generated by clustering top variable CpG-island probes \\nmapped to ± 1500bp of TSS that had a starting β value of ≤ 0.2 in controls at one month. \\nHeatmaps for hypomethylated probes were generated by clustering the top 2.5% variable \\nCpG-island probes mapped to ± 1500bp of TSS and had a β value of > 0.5 in the control \\ncells are one month.\\nDAVID analysis of methylated gene lists (GO analysis_biological processes)—\\nProbes with a Δ β cut off of ≥ 0.3 picked from methylated gene list obtained as described \\nabove, from replicate 1 and 2 (Table 1) were combined and analyzed for Gene Ontology \\n(GO) enrichment for biological processes (BP) using the DA VID Bioinformatics resources \\ndatabase ( Huang da et al., 2009a ; Huang da et al., 2009b ). Only categories that were below \\nthe false discovery rate (Benjamini and Hochberg) of 0.05 are reported. The methylated gene \\nlist (Table S1) was also analyzed using the DA VID Functional Annotation Tool (version 6.7) \\nto obtain enriched functional categories at each time point and a list of genes enriched for in \\na particular functional category was generated as reported in Table S2.\\nComparing methylated gene lists with TCGA data (used to generate venn \\ndiagrams)— Infinium methylation array data for primary NSCLC, breast, colon and \\npancreatic tumors were used to test if genes identified as methylated in CSC treated cells are \\nalso methylated in primary tumors. Data were downloaded from the Cancer Genome Atlas \\n(TCGA; http://tcga.cancer.gov/dataportal ). β values were calculated as above followed by \\nquantile normalization of samples within the same tumor type. A conservative cut-off of β \\nvalues > 0.5 in the TCGA data was used to call samples as methylated. Hypermethylated \\nprobes in LUAD and LUSC were identified as follows: For each probe that have mean β \\nvalue in normal cells ≤ 0.2, we estimated the proportion of samples in the tumors that have β \\nvalue greater than 2 standard deviations from the mean β value of normal and that the \\ndifference in β value for the probe in the sample relative to the mean β value of normal is \\ngreater than 0.2. Those probes that are methylated in at least 25 % of the tumor samples Vaz et al. Page 14\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript\", metadata={'source': 'cancercell.pdf', 'page': 13}),\n",
      " Document(page_content='were defined as hypermethylated and the corresponding genes were identified as described \\nearlier.\\nChromatin immunoprecipitation (ChIP) and ChIP-seq analysis— For ChIP and \\nChIP-seq studies, cells were crosslinked using 1% formaldehyde. Nuclei were isolated using \\nCEBN buffer as described above, lysed and the chromatin was sonicated to obtain fragments \\nranging from 150–250 bp. Chromatin was immunoprecipitated by antibodies to \\nDNMT1(Sigma), EZH2 XP (Cell Signaling), H3K4Me3 (Millipore) and H3K27Me3 \\n(Millipore) and captured on ProteinA/G magnetic beads (DynaBeads). Bound chromatin was \\nwashed in low-salt, high-salt buffers and TE and subjected to elution, de-crosslinking, and \\nProteinase-K treatment at 65°C in elution buffer followed by col umn purification (Qiagen, \\nMin Elute). Primers used for ChIP are indicated in Table S4. ChIP was done with cells from \\nboth replicates at all time points indicated.\\nFor ChIP-seq studies purified DNA was then used to prepare a sequencing library and \\nsequenced on Applied Biosystems SOLiD (V3). Sequencing reads were aligned to hg19 \\n(UCSC) using bowtie1.0.0, which guarantees finding all alignments between the first 25 bp \\nof the read (seed) and the reference sequence with up to two mismatches. Each match is \\nextended to the full length of the read, scoring 1 point for matching and −2 points for \\nmismatching bases. The read is trimmed to the length with the highest score. If there is only \\none alignment or if an alignment scores significantly higher than the others for the same \\nread, it is considered unique and reported. ChIP-seq reads were normalized to total read \\ncounts in each sample. CSC methylated genes used for this analysis were picked from the 10 \\nmonth methylated gene list from replicate 1 since ChIP-seq analysis was performed with \\ncells obtained from this replicate. The methylated gene list was filtered to pick genes which \\nshowed Δ β Con (10 M-1 M) ≤ 0.2 and Δ β CSC 10 M - Con 10 M ≥ 0.35. A similar number \\nof random CpG-island containing but unmethylated genes were picked as a control set. For \\nthe analysis using heatmaps in Figure 4B, for the CSC methylated and a random set of \\nunmethylated genes, we obtained the ChIP-seq reads in 200bp bins (obtained using \\nBedTools) ± 5Kb around the TSS. The rows in the heatmap corresponding to the TSS region \\nwere ordered by lowest to highest (top to bottom) number of reads in the H3K4me3 ChIP-\\nseq in the control samples at 10 days. This order was maintained for all the other ChIP-seq \\ndata plotted in these heatmaps. Average plots were generated by averaging the normalized \\nChIP-seq reads per bin for the set of genes in each plot.\\nGene expression analysis— Gene expression was studied in the control and CSC treated \\ncells at 10 days, 1, 3, 6, 10 and 15 months as well as in the 15 month control and CSC \\ntreated cells expressing empty vector (EV) or KRASV12 and the tumor xenografts . These \\nsamples were collected from both replicates. RNA was extracted using the RNeasy Mini kit \\n(Qiagen) and processed for hybridization on an Human Gene Expression v2 4x44K \\nMicroarray chip. The R/Bioconductor package limma was used to process expression data. \\nWithin- and between-array normalizations were performed using the loess and aquantile \\nmethods, respectively. The normexp option was used for background correction. Raw files \\nwere read in using the read.maimages function. Log fold change in gene expression for CSC \\ntreated over respective controls was obtained for each sample at each time point studied. Vaz et al. Page 15\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 14}),\n",
      " Document(page_content='Ranked lists of log fold change were analyzed using Gene Set Enrichment Analysis (GSEA) \\nby the Broad Institute and data packages (Reactome, KEGG, oncogenic) ( Subramanian et \\nal., 2005 ). Pathways enriched with a false discovery rate less than 0.25 and p value < 0.05 \\nwere chosen. For analysis of gene expression in mouse xenografts obtained from replicate 1 \\nand 2, all genes that were upregulated/downregulated by 2 fold or higher as compared to the \\nrespective control cells with KRASG12V were selected. Ranked lists were prepared with \\nthese genes and analyzed by GSEA using the same data packages described above. Gene \\nexpression changes were also plotted for a key set of methylated genes identified in Table S2 \\n(Figure 7 and Figure S4). Fold change in gene expression of these genes in the tumor \\nxenografts was calculated by comparing with their expression levels in the respective control \\ncells expressing KRASV12.\\nqRT-PCR to validate expression of select genes analyzed in ChIP studies in control and CSC \\ntreated cells at the indicated time points was performed using the primers in Table S4.\\nOverexpression of KRAS12V in HBEC cells— Control and CSC treated cells at 6 and \\n15 months were transduced with Lentiviral vectors expressing mutant KRAS12V or an empty \\nvector control. Vectors were gifts from Daniel Haber (Addgene plasmid # 35633 and # \\n19066) ( Singh et al., 2012 ).Transduced cells were selected with hygromycin. Expression of \\nKRASV12 was confirmed by western blot using a KRAS G12V mutant specific antibody \\n(Cell Signaling).\\nDownregulation of TP53 with shRNA— The following shRNA sequence for TP-53 \\n(ACGGCGCACAGAGGAAGAGAAT) was cloned in the pRFP-CB-shLenti Lentiviral \\nvector (TR30032) by OriGene. This sequence was obtained from SIGMA mission shRNA \\nsequences shown to have on efficiency of knocking down TP53 by almost 90%. A control \\nscrambled shRNA cloned in the similar vector (TR30033) obtained from OriGene was used \\nas a negative control. Cells were transduced with lentiviral constructs expressing either \\nshTP53 or scrambled shRNA control and selected by growth in medium with Blasticidin S.\\nSoft agar growth assays— Clonogenic potential of control and CSC treated cells at 10 \\nmonths was determined by soft agar assays. 1,000 viable cells were suspended and plated in \\n0.4% agar in K-SFM medium in 6-well plates, and were layered over a 0.6% agar base in the \\nsame medium. Colonies were stained with ethidium bromide prior to enumeration. Similar \\nassays were also done to determine clonogenicity of 6 month and 15 month control and CSC \\ntreated cells expressing KRASV12.\\nIn vivo tumorigenicity and histologic analysis— 2 × 106 viable CSC and control \\ncells expressing KRASV12 in 0.2 mL of PBS were injected subcutaneously in the flank of \\nmale NSG mice (5–6 weeks). Mice were monitored every 3 days for tumor formation. \\nFormalin-fixed, paraffin-embedded (FFPE) xenograft tumor tissue was sectioned and stained \\nwith hematoxylin and eosin (H&E) and mucicarmine for histologic analysis. \\nImmunohistochemical staining for p63, TTF1, CK5 and CK7 was done in the Johns Hopkins \\nMedicine Department of Pathology Immunohistochemistry laboratory.Vaz et al. Page 16\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 15}),\n",
      " Document(page_content='Immunofluorescent Staining of Vimentin and E-cadherin— Cells grown on \\nchamber slides were fixed in 4% formaldehyde for 15 min, following which they were \\nwashed thrice in 1X PBS for 5 min each. Cells were then blocked for 1 hr in blocking buffer \\n(1X PBS/5% normal goat serum/0.3% Triton X-100). Primary anitbodies (E-cadherin and \\nVimentin) were added at 1.100 dilution overnight at 4°C. Slides were then washed thrice \\nwith 1X PBS and flurochrome-conjugated secondary antibodies added at 1:500 dilution for \\n1–2 hr at RT. Slides were then washed thrice with 1X PBS and mounted in Prolong Gold \\nAntifade with DAPI (Cell Signaling Technology).\\nSample preparation for whole-exome and targeted sequencing— Samples were \\nprepared, sequenced and analyzed as described earlier ( Bertotti et al., 2015 ). DNA was \\nextracted from cells and xenograft tissues using the Qiagen DNeasy blood and tissue mini \\nkit (Qiagen, CA). Fragmented genomic DNA from tumor and normal samples was used for \\nIllumina TruSeq library construction (Illumina, San Diego, CA). Exonic regions or targeted \\nregions (targeted genes-Table S3) were captured in solution using the Agilent SureSelect v.4 \\nkit (Agilent, Santa Clara, CA) according to the manufacturer’s instructions as previously \\ndescribed ( Bertotti et al., 2015 ; Sausen et al., 2013 ). Paired-end sequencing, resulting in 100 \\nor 150 bases from each end of the DNA fragments for the exome or targeted sequencing \\nlibraries, respectively, was performed using Illumina HiSeq 2000/2500 instrumentation \\n(Illumina, San Diego, CA).\\nAnalysis of next-generation sequencing data and identification of putative \\nsomatic mutations— Genetic alterations were identified using VariantDx custom software \\nfor identifying mutations in samples by removing background mutations by using a specified \\ncontrol as previously described ( Bertotti et al., 2015 ; Sausen et al., 2013 ). Before mutation \\ncalling, primary processing of sequence data both the sample and for control was performed \\nusing Illumina CASA V A software (version 1.8), including masking of adaptor sequences. \\nSequence reads were aligned against the human reference genome (version hg19) using \\nELAND with additional realignment of select regions using the Needleman-Wunsch \\nmethod. Candidate mutations, consisting of point mutations, insertions, and deletions, were \\nthen identified using VariantDx across the whole exome. VariantDx examines sequence \\nalignments of samples against a control while applying filters to exclude alignment and \\nsequencing artifacts. Potential alterations were compared with mouse sequences from \\nexperimentally obtained mouse whole-exome and targeted sequence data as well as the \\nreference mouse genome (mm9) to remove mouse-specific variants. Amplification analyses \\nwere performed using the digital karyotyping approach ( Leary et al., 2007 ) by comparing \\nthe number of reads mapping to a particular gene with the average number of reads mapping \\nto each gene in the panel, along with a minor allele fraction analysis of heterozygous single \\nnucleotide polymorphisms contained within each gene. Candidate somatic mutations were \\nfurther filtered on the basis of gene annotation to identify those occurring in protein coding \\nregions. Functional consequences were predicted using snpEff and a custom database of \\nCCDS, RefSeq and Ensembl annotations using the latest transcript versions available on \\nhg19 from UCSC ( https://genome.ucsc.edu/ ). Predictions were ordered to prefer transcripts \\nwith canonical start and stop codons and CCDS or Refseq transcripts over Ensembl when \\navailable.Vaz et al. Page 17\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 16}),\n",
      " Document(page_content='Generation of box plots for analysis of methylation data from NSCLC samples \\nfrom TCGA for correlation with mutational and smoking status— Infinium \\nmethylation array data for primary lung adenocarcinoma and adjacent normal tissue was \\nused to examine the methylation status of the CSC methylated genes in these tumors and \\nnormal samples (depicted in Figures 7F and 7G) when separated by mutational status and/or \\nsmoking history. Mutational data for generating figures 7F and 7G was obtained from \\npreviously published data ( Cancer Genome Atlas Research, 2014 ) and methyaltion data for \\nthe same samples was obtained from the Cancer Genome Atlas (TCGA; http://\\ntcga.cancer.gov/dataportal ). A key set of CSC-methylated genes from Table S2 used earlier \\nfor analysis in Figures 7E and 7F were used to test the methylation status of these genes in \\nthe primary normal and tumor tissues. Box plots are made by plotting all the CpG-island \\nprobes mapped to ± 1500bp of TSS of these genes. Data are presented as the average of all \\nprobes in all patients within the respective group.\\nGeneration of box plots for analysis of methylation data from COPD (Chronic \\nlung obstructive disease) and IPF (idiopathic Pulmonary Fibrosis) samples—\\nMethylation data for analysis of COPD samples depicted in Figure S3B were obtained from \\nthe Lung Genomics Research consortium data portal ( http://www.lung-genomics.org ). \\nMethylation in these samples was generate using the CHARM platform. The normalized \\nCHARM values and the meta-data was downloaded from Lung Genomics Research \\nConsortium data portal. CHARM data was annotated using RefSeq Genes (NCBI36/gh18) \\nand Canonical CpG Islands (NCBI36/gh18) from UCSC Genome Bowser \\n(genome.ucsc.edu ). Box plots were generated by plotting all the CpG-island probes mapping \\nto ± 1500bp of TSS of 15month CSC-methylated genes from rep1 and rep2 (Table S1) in \\nCOPD and normal samples.\\nMethylation data for analysis of COPD, IPF and lung cancer samples depicted in Figure S3C \\nwere obtained from the GEO database the GEO database (GSE63704). Infinium 450K \\nmethylation data was available for these samples. Box plots were generated by averaging all \\nthe CpG-island probes mapped to ± 1500bp of TSS for each individual and plotting the \\naverage β values of promoter CpG-island probes. Python package seaborn was used for all \\nof the box plots analysis.\\nQUANTIFICATION AND STATISTICAL ANALYSIS\\nAll western blot and ChIP data are presented as the mean ± standard deviation (SD). These \\ndata are evaluated by two-tailed t test using Excel software. Box and whisker plots (Figures \\n7C, 7D, S4H and Figure S6) were generated in GraphPad Prism® (GraphPad Software, Inc.) \\nand analyzed using the Wilcoxon matched-pairs signed rank test. Box plots in Figure 7F, 7G \\nand S3b and S3C were generated using Python package seaborn and p values were \\ncalculated using Welch’s t-test. p values of < 0.05 were considered statistically significant \\nfor all tests. Where appropriate, figure legends define n, which indicates the number of \\nindividual biological replicates.Vaz et al. Page 18\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 17}),\n",
      " Document(page_content='DATA AVAILABILITY\\nAll microarray, methylation and ChIP-seq data has been deposited in GEO under the series \\nreference GSE101864.\\nSupplementary Material\\nRefer to Web version on PubMed Central for supplementary material.\\nAcknowledgments\\nResearch reported in this publication was supported by the National Cancer Institute under award number \\nR01CA043318 (S.B.B.), R01CA185357(H.E.), National Institutes of Health R01CA121113 (V .E.V .), \\nP30CA006973 (S.B.B.) (V .E.V .), U10CA180950 (V .E.V), NIEHS, R01ES011858 (S.B.B), R01ES023183 \\n(H.M.O’H). The content is solely the responsibility of the authors and does not necessarily represent the official \\nviews of the National Institutes of Health. The Hodson Trust (S.B.B.), the Commonwealth Foundation (V .E.V .). The \\nDr. Miriam and Sheldon G. Adelson Medical Research Foundation (V .E.V ., S.B.B.). SU2C-DCS International \\nTranslational Cancer Research Dream Team (SU2C-AACR-DT1415; V .E.V .), Stand Up To Cancer is a program of \\nthe Entertainment Industry Foundation, Administered by AACR. Flight Attendant Medical Research Institute \\n(M.V .). We thank Kathy Bender for help with manuscript preparation.\\nReferences\\nAran D, Sabato S, Hellman A. DNA methylation of distal regulatory sites characterizes dysregulation \\nof cancer genes. Genome Biol. 2013; 14:R21. [PubMed: 23497655] \\nAzuara V , Perry P, Sauer S, Spivakov M, Jorgensen HF, John RM, Gouti M, Casanova M, Warnes G, \\nMerkenschlager M, Fisher AG. Chromatin signatures of pluripotent cell lines. Nature cell biology. \\n2006; 8:532–538. [PubMed: 16570078] \\nBaylin SB. The cancer epigenome: its origins, contributions to tumorigenesis, and translational \\nimplications. Proceedings of the American Thoracic Society. 2012; 9:64–65. [PubMed: 22550245] \\nBaylin SB, Jones PA. A decade of exploring the cancer epigenome - biological and translational \\nimplications. Nature reviews Cancer. 2011; 11:726–734. [PubMed: 21941284] \\nBaylin SB, Ohm JE. Epigenetic gene silencing in cancer - a mechanism for early oncogenic pathway \\naddiction? Nature reviews Cancer. 2006; 6:107–116. [PubMed: 16491070] \\nBelinsky SA, Palmisano WA, Gilliland FD, Crooks LA, Divine KK, Winters SA, Grimes MJ, Harms \\nHJ, Tellez CS, Smith TM, et al. Aberrant promoter methylation in bronchial epithelium and sputum \\nfrom current and former smokers. Cancer research. 2002; 62:2370–2377. [PubMed: 11956099] \\nBernstein BE, Mikkelsen TS, Xie X, Kamal M, Huebert DJ, Cuff J, Fry B, Meissner A, Wernig M, \\nPlath K, et al. A bivalent chromatin structure marks key developmental genes in embryonic stem \\ncells. Cell. 2006; 125:315–326. [PubMed: 16630819] \\nBertotti A, Papp E, Jones S, Adleff V , Anagnostou V , Lupo B, Sausen M, Phallen J, Hruban CA, \\nTokheim C, et al. The genomic landscape of response to EGFR blockade in colorectal cancer. \\nNature. 2015; 526:263–267. [PubMed: 26416732] \\nBibikova M, Fan JB. GoldenGate assay for DNA methylation profiling. Methods Mol Biol. 2009; \\n507:149–163. [PubMed: 18987813] \\nBoch C, Kollmeier J, Roth A, Stephan-Falkenau S, Misch D, Gruning W, Bauer TT, Mairinger T. The \\nfrequency of EGFR and KRAS mutations in non-small cell lung cancer (NSCLC): routine \\nscreening data for central Europe from a cohort study. BMJ open. 2013; 3\\nCamilo R, Capelozzi VL, Siqueira SA, Del Carlo Bernardi F. Expression of p63, keratin 5/6, keratin 7, \\nand surfactant-A in non-small cell lung carcinomas. Human pathology. 2006; 37:542–546. \\n[PubMed: 16647951] \\nCancer Genome Atlas Research, N. Comprehensive genomic characterization of squamous cell lung \\ncancers. Nature. 2012; 489:519–525. [PubMed: 22960745] \\nCancer Genome Atlas Research, N. Comprehensive molecular profiling of lung adenocarcinoma. \\nNature. 2014; 511:543–550. [PubMed: 25079552] Vaz et al. Page 19\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 18}),\n",
      " Document(page_content=\"Clements EG, Mohammad HP, Leadem BR, Easwaran H, Cai Y , Van Neste L, Baylin SB. DNMT1 \\nmodulates gene expression without its catalytic activity partially through its interactions with \\nhistone-modifying enzymes. Nucleic acids research. 2012; 40:4334–4346. [PubMed: 22278882] \\nCowling VH, D'Cruz CM, Chodosh LA, Cole MD. c-Myc transforms human mammary epithelial cells \\nthrough repression of the Wnt inhibitors DKK1 and SFRP1. Molecular and cellular biology. 2007; \\n27:5135–5146. [PubMed: 17485441] \\nDai Z, Lakshmanan RR, Zhu WG, Smiraglia DJ, Rush LJ, Fruhwald MC, Brena RM, Li B, Wright FA, \\nRoss P, et al. Global methylation profiling of lung cancer identifies novel methylated genes. \\nNeoplasia. 2001; 3:314–323. [PubMed: 11571631] \\nDamiani LA, Yingling CM, Leng S, Romo PE, Nakamura J, Belinsky SA. Carcinogen-induced gene \\npromoter hypermethylation is mediated by DNMT1 and causal for transformation of immortalized \\nbronchial epithelial cells. Cancer research. 2008; 68:9005–9014. [PubMed: 18974146] \\nDelgado O, Kaisani AA, Spinola M, Xie XJ, Batten KG, Minna JD, Wright WE, Shay JW. Multipotent \\ncapacity of immortalized human bronchial epithelial cells. PloS one. 2011; 6:e22023. [PubMed: \\n21760947] \\nDimauro T, David G. Ras-induced senescence and its physiological relevance in cancer. Current cancer \\ndrug targets. 2010; 10:869–876. [PubMed: 20718709] \\nDurinck S, Spellman PT, Birney E, Huber W. Mapping identifiers for the integration of genomic \\ndatasets with the R/Bioconductor package biomaRt. Nature protocols. 2009; 4:1184–1191. \\n[PubMed: 19617889] \\nEaswaran H, Johnstone SE, Van Neste L, Ohm J, Mosbruger T, Wang Q, Aryee MJ, Joyce P, Ahuja N, \\nWeisenberger D, et al. A DNA hypermethylation module for the stem/progenitor cell signature of \\ncancer. Genome research. 2012; 22:837–849. [PubMed: 22391556] \\nEaswaran H, Tsai HC, Baylin SB. Cancer epigenetics: tumor heterogeneity, plasticity of stem-like \\nstates, and drug resistance. Molecular cell. 2014; 54:716–727. [PubMed: 24905005] \\nFuks F, Burgers WA, Brehm A, Hughes-Davies L, Kouzarides T. DNA methyltransferase Dnmt1 \\nassociates with histone deacetylase activity. Nature genetics. 2000; 24:88–91. [PubMed: \\n10615135] \\nGovindan R, Ding L, Griffith M, Subramanian J, Dees ND, Kanchi KL, Maher CA, Fulton R, Fulton \\nL, Wallis J, et al. Genomic landscape of non-small cell lung cancer in smokers and never-smokers. \\nCell. 2012; 150:1121–1134. [PubMed: 22980976] \\nHa K, Lee GE, Palii SS, Brown KD, Takeda Y , Liu K, Bhalla KN, Robertson KD. Rapid and transient \\nrecruitment of DNMT1 to DNA double-strand breaks is mediated by its interaction with multiple \\ncomponents of the DNA damage response machinery. Human molecular genetics. 2011; 20:126–\\n140. [PubMed: 20940144] \\nHelman E, Naxerova K, Kohane IS. DNA hypermethylation in lung cancer is targeted at \\ndifferentiation-associated genes. Oncogene. 2012; 31:1181–1188. [PubMed: 21804601] \\nHuang da W, Sherman BT, Lempicki RA. Bioinformatics enrichment tools: paths toward the \\ncomprehensive functional analysis of large gene lists. Nucleic acids research. 2009a; 37:1–13. \\n[PubMed: 19033363] \\nHuang da W, Sherman BT, Lempicki RA. Systematic and integrative analysis of large gene lists using \\nDA VID bioinformatics resources. Nature protocols. 2009b; 4:44–57. [PubMed: 19131956] \\nIssa JP. Aging and epigenetic drift: a vicious cycle. The Journal of clinical investigation. 2014; 124:24–\\n29. [PubMed: 24382386] \\nJones PA, Baylin SB. The epigenomics of cancer. Cell. 2007; 128:683–692. [PubMed: 17320506] \\nKaufman CK, Mosimann C, Fan ZP, Yang S, Thomas AJ, Ablain J, Tan JL, Fogley RD, van Rooijen E, \\nHagedorn EJ, et al. A zebrafish melanoma model reveals emergence of neural crest identity during \\nmelanoma initiation. Science. 2016; 351:aad2197. [PubMed: 26823433] \\nKraunz KS, Nelson HH, Liu M, Wiencke JK, Kelsey KT. Interaction between the bone morphogenetic \\nproteins and Ras/MAP-kinase signalling pathways in lung cancer. British journal of cancer. 2005; \\n93:949–952. [PubMed: 16175182] \\nLeary RJ, Cummins J, Wang TL, Velculescu VE. Digital karyotyping. Nat Protoc. 2007; 2:1973–1986. \\n[PubMed: 17703209] Vaz et al. Page 20\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript\", metadata={'source': 'cancercell.pdf', 'page': 19}),\n",
      " Document(page_content=\"Lee E, Yang J, Ku M, Kim NH, Park Y , Park CB, Suh JS, Park ES, Yook JI, Mills GB, et al. Metabolic \\nstress induces a Wnt-dependent cancer stem cell-like state transition. Cell death & disease. 2015; \\n6:e1805. [PubMed: 26136078] \\nLemjabbar-Alaoui H, Dasari V , Sidhu SS, Mengistab A, Finkbeiner W, Gallup M, Basbaum C. Wnt \\nand Hedgehog are critical mediators of cigarette smoke-induced lung cancer. PloS one. 2006; \\n1:e93. [PubMed: 17183725] \\nLiu F, Killian JK, Yang M, Walker RL, Hong JA, Zhang M, Davis S, Zhang Y , Hussain M, Xi S, et al. \\nEpigenomic alterations and gene expression profiles in respiratory epithelia exposed to cigarette \\nsmoke condensate. Oncogene. 2010; 29:3650–3664. [PubMed: 20440268] \\nMacaluso M, Paggi MG, Giordano A. Genetic and epigenetic alterations as hallmarks of the intricate \\nroad to cancer. Oncogene. 2003; 22:6472–6478. [PubMed: 14528270] \\nMaegawa S, Gough SM, Watanabe-Okochi N, Lu Y , Zhang N, Castoro RJ, Estecio MR, Jelinek J, \\nLiang S, Kitamura T, et al. Age-related epigenetic drift in the pathogenesis of MDS and AML. \\nGenome research. 2014; 24:580–591. [PubMed: 24414704] \\nMazieres J, He B, You L, Xu Z, Lee AY , Mikami I, Reguart N, Rosell R, McCormick F, Jablons DM. \\nWnt inhibitory factor-1 is silenced by promoter hypermethylation in human lung cancer. Cancer \\nresearch. 2004; 64:4717–4720. [PubMed: 15256437] \\nMohammad HP, Cai Y , McGarvey KM, Easwaran H, Van Neste L, Ohm JE, O'Hagan HM, Baylin SB. \\nPolycomb CBX7 promotes initiation of heritable repression of genes frequently silenced with \\ncancer-specific DNA hypermethylation. Cancer research. 2009; 69:6322–6330. [PubMed: \\n19602592] \\nNguyen DX, Chiang AC, Zhang XH, Kim JY , Kris MG, Ladanyi M, Gerald WL, Massague J. \\nWNT/TCF signaling through LEF1 and HOXB9 mediates lung adenocarcinoma metastasis. Cell. \\n2009; 138:51–62. [PubMed: 19576624] \\nNiwa T, Ushijima T. Induction of epigenetic alterations by chronic inflammation and its significance \\non carcinogenesis. Advances in genetics. 2010; 71:41–56. [PubMed: 20933125] \\nNojima M, Suzuki H, Toyota M, Watanabe Y , Maruyama R, Sasaki S, Sasaki Y , Mita H, Nishikawa N, \\nYamaguchi K, et al. Frequent epigenetic inactivation of SFRP genes and constitutive activation of \\nWnt signaling in gastric cancer. Oncogene. 2007; 26:4699–4713. [PubMed: 17297461] \\nO'Hagan HM, Mohammad HP, Baylin SB. Double strand breaks can initiate gene silencing and SIRT1-\\ndependent onset of DNA methylation in an exogenous promoter CpG island. PLoS genetics. 2008; \\n4:e1000155. [PubMed: 18704159] \\nO'Hagan HM, Wang W, Sen S, Destefano Shields C, Lee SS, Zhang YW, Clements EG, Cai Y , Van \\nNeste L, Easwaran H, et al. Oxidative damage targets complexes containing DNA \\nmethyltransferases, SIRT1, and polycomb members to promoter CpG Islands. Cancer cell. 2011; \\n20:606–619. [PubMed: 22094255] \\nOhm JE, Baylin SB. Stem cell chromatin patterns: an instructive mechanism for DNA \\nhypermethylation? Cell Cycle. 2007; 6:1040–1043. [PubMed: 17457052] \\nPacheco-Pinedo EC, Durham AC, Stewart KM, Goss AM, Lu MM, Demayo FJ, Morrisey EE. Wnt/\\nbeta-catenin signaling accelerates mouse lung tumorigenesis by imposing an embryonic distal \\nprogenitor phenotype on lung epithelium. The Journal of clinical investigation. 2011; 121:1935–\\n1945. [PubMed: 21490395] \\nPacheco-Pinedo EC, Morrisey EE. Wnt and Kras signaling-dark siblings in lung cancer. Oncotarget. \\n2011; 2:569–574. [PubMed: 21753228] \\nRamirez RD, Sheridan S, Girard L, Sato M, Kim Y , Pollack J, Peyton M, Zou Y , Kurie JM, Dimaio \\nJM, et al. Immortalization of human bronchial epithelial cells in the absence of viral oncoproteins. \\nCancer research. 2004; 64:9027–9034. [PubMed: 15604268] \\nRevet I, Huizenga G, Chan A, Koster J, V olckmann R, van Sluis P, Ora I, Versteeg R, Geerts D. The \\nMSX1 homeobox transcription factor is a downstream target of PHOX2B and activates the Delta-\\nNotch pathway in neuroblastoma. Experimental cell research. 2008; 314:707–719. [PubMed: \\n18201699] \\nRevet I, Huizenga G, Koster J, V olckmann R, van Sluis P, Versteeg R, Geerts D. MSX1 induces the \\nWnt pathway antagonist genes DKK1, DKK2, DKK3, and SFRP1 in neuroblastoma cells, but does Vaz et al. Page 21\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript\", metadata={'source': 'cancercell.pdf', 'page': 20}),\n",
      " Document(page_content='not block Wnt3 and Wnt5A signalling to DVL3. Cancer letters. 2010; 289:195–207. [PubMed: \\n19815336] \\nRountree MR, Bachman KE, Baylin SB. DNMT1 binds HDAC2 and a new co-repressor, DMAP1, to \\nform a complex at replication foci. Nature genetics. 2000; 25:269–277. [PubMed: 10888872] \\nSato M, Larsen JE, Lee W, Sun H, Shames DS, Dalvi MP, Ramirez RD, Tang H, DiMaio JM, Gao B, \\net al. Human lung epithelial cells progressed to malignancy through specific oncogenic \\nmanipulations. Molecular cancer research : MCR. 2013; 11:638–650. [PubMed: 23449933] \\nSato M, Vaughan MB, Girard L, Peyton M, Lee W, Shames DS, Ramirez RD, Sunaga N, Gazdar AF, \\nShay JW, Minna JD. Multiple oncogenic changes (K-RAS(V12), p53 knockdown, mutant EGFRs, \\np16 bypass, telomerase) are not sufficient to confer a full malignant phenotype on human \\nbronchial epithelial cells. Cancer research. 2006; 66:2116–2128. [PubMed: 16489012] \\nSausen M, Leary RJ, Jones S, Wu J, Reynolds CP, Liu X, Blackford A, Parmigiani G, Diaz LA Jr, \\nPapadopoulos N, et al. Integrated genomic analyses identify ARID1A and ARID1B alterations in \\nthe childhood cancer neuroblastoma. Nature genetics. 2013; 45:12–17. [PubMed: 23202128] \\nSchlesinger Y , Straussman R, Keshet I, Farkash S, Hecht M, Zimmerman J, Eden E, Yakhini Z, Ben-\\nShushan E, Reubinoff BE, et al. Polycomb-mediated methylation on Lys27 of histone H3 pre-\\nmarks genes for de novo methylation in cancer. Nature genetics. 2007; 39:232–236. [PubMed: \\n17200670] \\nShen H, Laird PW. Interplay between the cancer genome and epigenome. Cell. 2013; 153:38–55. \\n[PubMed: 23540689] \\nSingh A, Sweeney MF, Yu M, Burger A, Greninger P, Benes C, Haber DA, Settleman J. TAK1 \\ninhibition promotes apoptosis in KRAS-dependent colon cancers. Cell. 2012; 148:639–650. \\n[PubMed: 22341439] \\nStange DE, Koo BK, Huch M, Sibbel G, Basak O, Lyubimova A, Kujala P, Bartfeld S, Koster J, \\nGeahlen JH, et al. Differentiated Troy+ chief cells act as reserve stem cells to generate all lineages \\nof the stomach epithelium. Cell. 2013; 155:357–368. [PubMed: 24120136] \\nStewart DJ. Wnt signaling pathway in non-small cell lung cancer. Journal of the National Cancer \\nInstitute. 2014; 106:djt356. [PubMed: 24309006] \\nSubramanian A, Tamayo P, Mootha VK, Mukherjee S, Ebert BL, Gillette MA, Paulovich A, Pomeroy \\nSL, Golub TR, Lander ES, Mesirov JP. Gene set enrichment analysis: a knowledge-based approach \\nfor interpreting genome-wide expression profiles. Proc Natl Acad Sci U S A. 2005; 102:15545–\\n15550. [PubMed: 16199517] \\nSuzuki H, Watkins DN, Jair KW, Schuebel KE, Markowitz SD, Chen WD, Pretlow TP, Yang B, \\nAkiyama Y , Van Engeland M, et al. Epigenetic inactivation of SFRP genes allows constitutive \\nWNT signaling in colorectal cancer. Nature genetics. 2004; 36:417–422. [PubMed: 15034581] \\nTata PR, Mou H, Pardo-Saganta A, Zhao R, Prabhu M, Law BM, Vinarsky V , Cho JL, Breton S, Sahay \\nA, et al. Dedifferentiation of committed epithelial cells into stem cells in vivo. Nature. 2013; \\n503:218–223. [PubMed: 24196716] \\nTellez CS, Juri DE, Do K, Bernauer AM, Thomas CL, Damiani LA, Tessema M, Leng S, Belinsky SA. \\nEMT and stem cell-like properties associated with miR-205 and miR-200 epigenetic silencing are \\nearly manifestations during carcinogen-induced transformation of human lung epithelial cells. \\nCancer research. 2011; 71:3087–3097. [PubMed: 21363915] \\nTessema M, Yingling CM, Liu Y , Tellez CS, Van Neste L, Baylin SS, Belinsky SA. Genome-wide \\nunmasking of epigenetically silenced genes in lung adenocarcinoma from smokers and never \\nsmokers. Carcinogenesis. 2014; 35:1248–1257. [PubMed: 24398667] \\nTorre LA, Bray F, Siegel RL, Ferlay J, Lortet-Tieulent J, Jemal A. Global cancer statistics, 2012. CA: a \\ncancer journal for clinicians. 2015; 65:87–108. [PubMed: 25651787] \\nWidschwendter M, Fiegl H, Egle D, Mueller-Holzner E, Spizzo G, Marth C, Weisenberger DJ, \\nCampan M, Young J, Jacobs I, Laird PW. Epigenetic stem cell signature in cancer. Nature genetics. \\n2007; 39:157–158. [PubMed: 17200673] \\nYou JS, Jones PA. Cancer genetics and epigenetics: two sides of the same coin? Cancer cell. 2012; \\n22:9–20. [PubMed: 22789535] Vaz et al. Page 22\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 21}),\n",
      " Document(page_content='SIGNIFICANCE\\nCancer development is a complex process involving both genetic and epigenetic \\nabnormalities. However, the evolution of these changes have not been well delineated. \\nWe demonstrate that exposure of human bronchial epithelial cells to chronic cigarette \\nsmoke causes progressive gene promoter DNA hypermethylation that associates with \\nactivation of key signaling pathways driving the tumorigenic process. The methylated \\ngene abnormalities highly overlap with those commonly seen in non-small cell lung \\ncancer. Most importantly, these epigenetic changes may predispose cells to single step \\ntransformation mediated by a single oncogene, mutated KRAS , in the absence of other \\ndriver genetic changes. This study provides a paradigm in which epigenetic alterations \\nmay precede and sensitize for genetic events to drive the development of lung cancer.Vaz et al. Page 23\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 22}),\n",
      " Document(page_content='HIGHLIGHTS\\nChronic smoke exposure causes sequential chromatin changes leading to gene silencing. \\nSilenced genes are normally polycomb controlled but adopt abnormal DNA methylation. \\nGene Methylation causes sequential upregulation of key signal transduction pathways. \\nEpigenetic alterations sensitize cells to transformation by a single oncogenic event.Vaz et al. Page 24\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 23}),\n",
      " Document(page_content='Figure 1. Expression of DNMT1, EZH2 and SIRT1 following chronic exposure to CSC\\n(A and B) Immunoblots showing expression of DNMT1, SIRT1 and EZH2 in nuclear \\nextracts (A) and tight chromatin fractions (B) obtained from cells treated with DMSO (Con), \\n5 µg/ml and 10 µg/ml CSC over a period of 15 months collected at the indicated time points. \\n(C) Quantitation of protein bands using ImageJ. The data presented is the mean of three \\nindependent experiments ± SD. * p < 0.05 by two tailed t-test. D) Phase contrast \\nphotomicrographs showing morphological changes observed in Control and 10 µg/ml CSC \\ntreated HBECs over time. Scale bar = 100 µm. See also Figure S1.Vaz et al. Page 25\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 24}),\n",
      " Document(page_content='Figure 2. Progressive changes in methylation in control and CSC treated cells\\n(A) Principal component analysis for control (Con) and 10 µg/ml CSC exposed cells of the \\nDNA methylation patterns, assayed by Illumina 450K arrays, for most variable probes at \\nexamined time points from two replicates in each treatment group. Green (Con) and orange \\n(CSC) circles indicate groupings. (B) Comparison of genome-wide DNA methylation \\nchanges, from Illumina 450K arrays, at promoter CpG islands (CpGi) between replicates of \\nthe same treatment group. (C) Comparison of genome-wide DNA methylation, obtained \\nfrom Illumina 450K arrays, at CpG islands in proximal promoter regions in control (x axis) \\nversus CSC (y axis) treated cells across the indicated time points in two replicates. (D) Vaz et al. Page 26\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 25}),\n",
      " Document(page_content='Scatter plots showing changes with time in methylation at promoter, gene body and \\nenhancer regions. Changes in β values for all CpG island probes within the respective \\nregions from 1 month (1 M) to the respective end point (6 M, 10 M and 15 M) are shown in \\ncontrol (X-axis) and CSC treated (Y-axis) samples. Red dots: probes with a β value < 0.2 in \\ncontrol cells at 1 month; blue dots: probes with β value > 0.5 in control cells at 1 month. Rep \\n1: replicate 1; Rep2: replicate 2. See also Figure S2.Vaz et al. Page 27\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 26}),\n",
      " Document(page_content='Figure 3. Identification and characterization of genes methylated upon CSC exposure\\n(A) Heat maps of β values for infinium probes in CpG islands within ± 1500 bp from TSS in \\ncontrol (Con) and 10 µg/ml CSC exposed cells from 1 to 15 months in replicates 1 and 2. \\nOnly probes with a starting β value of < 0.2 in control cells at 1 month were selected. Three \\nclusters of probes are identified; group I: methylated in controls and further methylated in \\nCSC treated cells; Group II: methylated only in CSC treated cells; Group III: methylated \\nonly in controls. (B) Top enriched biological processes in genes methylated in CSC treated \\ncells at 10 and 15 months of treatment. (C) Venn diagrams showing overlap of methylated \\ngenes in 10 µg/ml CSC treated cells at 10 and 15 months with genes hypermethylated in Vaz et al. Page 28\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 27}),\n",
      " Document(page_content='lung adenocarcinoma (LUAD) and lung squamous cell carcinomas (LUSC). To obtain genes \\nhypermethylated in LUAD and LUSC we picked probes methylated in at least 25% of tumor \\nsamples and identified corresponding genes as described in STAR methods. See also Figure \\nS3, Tables S1 and S2. M: months.Vaz et al. Page 29\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 28}),\n",
      " Document(page_content='Figure 4. Progressive changes in chromatin marks and binding of chromatin-associated proteins \\nin CSC exposed cells\\n(A) ChIP studies examining enrichment of H3K4me3, H3K27me3 and binding of EZH2 and \\nDNMT1 at representative genes in control and CSC treated cells over time. Cells were \\ntreated with DMSO or 10 µg/ml CSC for the indicated time periods before processing for \\nChIP. Data are presented as mean ± SD (n = 3). * p < 0.05 by two tailed t-test. (B) Gene \\nexpression of representative genes at the indicated time points following treatment with 10 \\nµg/mL CSC. Values indicate fold change in CSC treated over control cells. Data are \\npresented as mean ± SD (n = 3). * p < 0.05 by two tailed t-test. (C) Genome wide ChIP-seq \\nanalysis in control (C) and CSC treated cells at 10 days (10 D) and 10 months (10 M) post Vaz et al. Page 30\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 29}),\n",
      " Document(page_content='onset of CSC exposure showing enrichment of EZH2, DNMT1, H3K4me3 and H3K27me3 \\nat promoters of genes methylated in CSC treated cells (CSC Methylated Genes) and a \\nrandom set of unmethylated CpG island containing genes (Random Unmethylated Genes). \\nX-axis; 0 = transcription start sites. For each mark analyzed, a graph showing quantitative \\ndata depicting the average read counts ± 5 kb from the transcription start site (TSS) for all \\ngenes in the respective groups is shown above the maps of read intensities for individual \\ngenes at 10 D and 10 M in C and CSC cells. See also Figure S4.Vaz et al. Page 31\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 30}),\n",
      " Document(page_content='Figure 5. Gene set enrichment analysis (GSEA) of genome-wide gene expression following CSC \\nexposure\\n(A) Genome wide gene expression analysis of control and 10 µg/ml CSC treated cells at the \\ntime points indicated over a period of 15 months. Ranked list of fold changes in gene \\nexpression in CSC treated cells as compared to matched controls were calculated and \\nanalyzed using GSEA. The top 20 enriched Pathways were compared across the time points. \\nHeatmap showing changes in enrichment (based on NES scores) of key pathways at various \\ntimes following CSC exposure. (B) Gene sets enriched in CSC treated cells at the indicated \\ntime points. Log fold change in gene expression in CSC over control cells was determined Vaz et al. Page 32\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 31}),\n",
      " Document(page_content='and ranked lists were analyzed using GSEA (FDR < 0.25, p value < 0.01). See also Figure \\nS5.Vaz et al. Page 33\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 32}),\n",
      " Document(page_content='Figure 6. Phenotypes of CSC exposed versus control cells before and after expression of mutant \\nKRASV12\\n(A) Soft agar assays to assess the anchorage-independent growth ability of HBEC cells \\nfollowing 10 months of CSC exposure. (B) Quantitation of soft agar colonies in panel A. \\nValues indicate average number of colonies ± SD from replicate 1 and 2 at the indicated time \\npoints. (C) Phase contrast photomicrographs showing morphological changes in control and \\nCSC treated cells when switched from defined serum-free to serum-containing medium for \\n96 hr. Scale bar = 50 µm. (D) Immunofluorescent staining of E-cadherin and Vimentin in \\ncells grown in serum-containing medium for 96 hr. Scale bar = 100 µm. (E) Immunoblot for \\nEMT markers in cells (CSC and Con) and soft agar clones obtained from 15 month CSC Vaz et al. Page 34\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 33}),\n",
      " Document(page_content='treated cells, grown in serum-containing medium for 96 hr. (F and G) Expression of \\nKRASV12 (F) and TP53 (G) in 15 month control and CSC treated cells two weeks following \\ntransduction with a KRAS V12 or shTP53 expressing lentiviral vector, respectively. Stably \\ninfected cells were selected by growth in the respective antibiotic prior to examining gene \\nexpression. (H) Soft agar assays to assess the clonogenic potential of 6 and 15 month control \\nand CSC treated cells expressing mutant KRASV12 (top panel). Representative images of \\nsoft agar colonies (bottom panel). Scale bar = 100 µm (I) Quantitation of colonies obtained \\nin H. Data are presented as average number of colonies ± SD at the indicated time points. * \\np < 0.05 by two tailed t-test (J) Signaling pathways enriched for in 15 month CSC treated \\ncells expressing KRASV12 compared to respective empty vector controls, analyzed using \\nGSEA. (K and L) Measurement of tumor latency (K) and average tumor volumes (L) of \\nxenografts from mice injected with 15 month control and CSC treated cells from replicate 1 \\n(Rep 1) and replicate 2 (Rep 2) expressing KRASV12 or shTP53. Data are presented as mean \\n± SD (n = 5 for Rep1 and n = 10 for Rep 2).Vaz et al. Page 35\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 34}),\n",
      " Document(page_content='Figure 7. Transformation of CSC exposed cells following expression of mutant KRASV12\\n(A) Expression of mutant KRASV12 in xenograft tissues obtained from mice injected with \\nCSC treated cells expressing KRASV12 (T1-T3, Rep 1 and T4-T6, Rep 2). 15 month CSC \\nexposed cells that were infected with empty vector (EV) are shown as a control. (B) \\nRepresentative histology of xenografts obtained from mice injected with CSC + KRASV12 \\ncell four weeks after appearance of the tumor. 5 sections were stained from each Replicate. \\nRepresentative sections stained with H&E, mucicarmine, CK5/6, CK7, and p63 are shown. \\nScale bar = 100 µm. (C) Comparison of methylation changes in key CSC-methylated genes \\nfrom Table S2 in empty vector (EV) and KRASV12 expressing control and CSC treated cells Vaz et al. Page 36\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 35}),\n",
      " Document(page_content='and xenografts (T1-T4) harvested 4 weeks after appearance of tumors from replicate 2. \\nBoxplots generated with β values of promoter CpG island probes of selected genes. Only \\nprobes with a starting β value of ≤ 0.4 in EV containing control cells were selected. \\nBoxplots represent the 25th and 75th percentiles, with midlines indicating the median \\nvalues. Whiskers extend to the minimum/maximum values of the data sample. * p < 0.01; ** \\np < 0.0001 by Wilcoxon rank test (D) Box plots showing average log2 fold change in \\nexpression of CSC-methylated genes from Table S2 in EV and KRASV12 expressing CSC \\nexposed cells and xenografts (T1-T4) from replicate 2 as compared to their respective \\ncontrol cells expressing EV or KRASV12. Boxplots represent the 25th and 75th percentiles, \\nwith midlines indicating the median values. Whiskers extend to the minimum/maximum \\nvalues of the data sample. ** p < 0.0001 by Wilcoxon rank test (E) Signaling pathways \\nenriched for in xenografts compared to control cells expressing KRASV12 analyzed with \\nGSEA. T: xenografts. (F) Box plots of β values of promoter CpG island probes of CSC \\nmethylated genes in tumors with KRAS  (n = 49), TP53  (n = 49) and EGFR  (n = 6) \\nmutations and corresponding matched normal tissues (n = 4 for KRAS , n =8 for TP53 ). Data \\nare presented as the average of all probes in all patients within the respective group. * p < \\n0.05 by Welch’s t-test indicates significant increase in matched normal from KRAS  tumors \\ncompared to matched normals from TP53  tumors. # p < 0.05 by Welch’s t-test indicates \\nsignificant decrease in EGFR mutant tumors compared to tumors with KRAS  and/or TP53 \\nmutation. (G) Box plots of β values of promoter CpG island probes of CSC methylated \\ngenes in tumors with KRAS  and TP53  mutations categorized by smoking status. * p < 0.05 \\nby Welch’s t-test indicates significant increase compared to non-smoker tumor within \\nmutation group. # p < 0.05 by Welch’s t-test indicates significant decrease compared to non-\\nsmoker tumor within mutation group. CRS: Current reformed smoker. (n = 8 current \\nsmokers, 19 CRS < 15 years, 14 CRS > 15 years, 4 non smokers for KRAS  tumors and n = \\n14 current smokers, 20 CRS < 15 years, 12 CRS > 15 years, 3 non smokers for TP53 \\ntumors). Top and bottom of Box plots shown in (F) and (G) represent the 25th and 75th \\npercentiles (interquartile range; IQR), with midlines indicating the median values. The \\nwhiskers extend to points that lie within 1.5 IQR’s of the lower and upper quartile. \\nObservations that fall outside this range are displayed independently. See also Figure S6.Vaz et al. Page 37\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.\\nAuthor Manuscript Author Manuscript Author Manuscript Author Manuscript', metadata={'source': 'cancercell.pdf', 'page': 36}),\n",
      " Document(page_content='Author Manuscript Author Manuscript Author Manuscript Author ManuscriptVaz et al. Page 38\\nTable 1\\nCandidate genes with a driver function\\nGene Name Function\\nBMP3 Member of the transforming growth factor-beta (TGF β) superfamily and represses KRAS/MAPK signaling in lung cancer. \\nInactivation is an early and frequent event in cancer development.\\nWIF1 Suppresses p38MAPK signaling, inhibits WNT pathway activation, suppresses stemness and induces cellular senescence in \\nsalivary gland tumors.\\nMSX1 Essential for apoptotic function of intact p53. Induces the WNT pathway antagonist genes DKK1 , DKK2 , DKK3 , and SFRP1 \\nin neuroblastoma cells\\nSFRP2 Negative regulator of WNT signaling.\\nSFRP5 Negative regulator of WNT signaling.\\nGATA4 Zinc-finger transcription factor that regulate genes involved in embryogenesis. Involved in GPCR signaling pathway and is \\nregulated by MAPK/ERK signaling.\\nSLIT2 Attenuated during lung cancer progression. Deregulates beta-catenin and E-cadherin and is associated with poor prognosis.\\nGALR1 Methylated in lung cancer. Identified as a tumor suppressor.\\nTable 1, See also Table S2: Candidate CSC-methylated genes with a driver function for the signaling events modulated by CSC treatment and \\nwhich are frequently methylated in lung cancer.\\nCancer Cell . Author manuscript; available in PMC 2018 September 11.', metadata={'source': 'cancercell.pdf', 'page': 37})]\n"
     ]
    }
   ],
   "source": [
    "from langchain_community.document_loaders import PyPDFLoader, OnlinePDFLoader\n",
    "from pprint import pprint\n",
    "\n",
    "loader=PyPDFLoader(\"cancercell.pdf\")\n",
    "# loader=OnlinePDFLoader(\"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5596892/pdf/nihms900592.pdf\")\n",
    "pdf_docs=loader.load()\n",
    "print(len(pdf_docs))\n",
    "pprint(pdf_docs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### web-based loader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Document(page_content='Hybrid Search ExplainedJanuary 3, 2023 · 7 min readErika CardenasTechnology Partner ManagerHybrid search is a technique that combines multiple search algorithms to improve the accuracy and relevance of search results. It uses the best features of both keyword-based search algorithms with vector search techniques. By leveraging the strengths of different algorithms, it provides a more effective search experience for users.The Hybrid search feature was introduced in Weaviate 1.17. It uses sparse and dense vectors to represent the semantic meaning and context of search queries and documents.In this blog post, you will learn about the implementation of hybrid search in Weaviate and how to use it.Sparse and dense vectors are calculated with distinct algorithms. Sparse vectors have mostly zero values with only a few non-zero values, while dense vectors mostly contain non-zero values. Sparse embeddings are generated from algorithms like BM25 and SPLADE. Dense embeddings are generated from machine learning models like\\nGloVe and Transformers.Note, the current implementation of hybrid search in Weaviate uses BM25/BM25F and vector search.If you’re interested to learn about how dense vector indexes are built and optimized in Weaviate, check out this article.BM25 builds on the keyword scoring method TF-IDF (Term-Frequency Inverse-Document Frequency) by taking the Binary Independence Model from the IDF calculation and adding a normalization penalty that weighs a document’s length relative to the average length of all the documents in the database.The image below presents the scoring calculation of BM25:\\n The score of the document, query pair is determined by weighing the uniqueness of each keyword in the query relative to the collection of texts. BM25 contains additional static parameters, k1 and b that may help calibrate performance to particular datasets.BM25F was also implemented in Weaviate 1.17. BM25F is a variant of BM25 that allows multiple text fields per object to be given different weights in the ranking calculation. These weights are important for when fields in a document are more important than others. For example, a title may be given more weight than the abstract, since the title is sometimes more informative and concise. This type of weighting makes BM25F more flexible and customizable than BM25.Dense vectors represent information stored in a database; this includes text, images, and other types of data. These embeddings are generated from machine learning models that convert data to vectors. The vectors are densely packed with information and are mostly made up of non-zero values. The meaning of each value in the vector depends on the machine learning model that you used.Vector databases, like Weaviate, store these embeddings and calculate the distance between the two vectors. Distance metrics show how similar or dissimilar two vector embeddings are. The search query is converted to a vector, similar to the data vectors, and the distance value determines how close the vectors are.Hybrid search merges dense and sparse vectors together to deliver the best of both search methods. Generally speaking, dense vectors excel at understanding the context of the query, whereas sparse vectors excel at keyword matches. Consider the query, “How to catch an Alaskan Pollock”. The dense vector representation is able to disambiguate “catch” as meaning fishing rather than baseball or sickness. The sparse vector search will match the phrase “Alaskan Pollock” only. This example query shows where hybrid search combines the best of both sparse and dense vectors.While researching hybrid search, we needed a way to combine the results of BM25 and dense vector search into a single ranked list. We came across a paper from Benham and Culpepper exploring rank fusion techniques. This paper analyzed seven strategies for combining the ranked results of two lists into a single ranking. We decided to start with the RRF score. The RRF score is calculated by taking the sum of the reciprocal rankings that is given from each list. By putting the rank of the document in the denominator, it penalizes the documents that are ranked lower in the list.Let\\'s look at an example of this. We have three documents labeled A, B, and C and have run a BM25 and Dense search. In this example, we have set the constant k to 0.The above table shows the ranked order of the BM25 and Dense search. To fuse the two lists together, we need to take the sum of the reciprocal ranks. Based on the results, the top document is Document B with a ranking of 1.5, then Document A at 1.3, and Document C at 0.83.Re-ranking is a necessary step when implementing hybrid search. The alpha parameter dictates the weighting of each algorithm and determines the re-ranking of the results.For a deep dive into how Weaviate\\'s hybrid search combines sparse and dense vectors, you can read more on these fusion algorithms in our blog post.Hybrid search queries are ideal for a search system that wants to leverage the power of semantic search capabilities but still rely on exact keyword matches. For example, the example search query \"How to catch an Alaskan Pollock\" from before would lead to better results with a hybrid search query than with a common keyword search or even a semantic search query.To use hybrid search in Weaviate, you only need to confirm that you’re using Weaviate v1.17 or a later version. You can run the hybrid queries in GraphQL or the other various client programming languages.There are five parameters needed to run the hybrid search query (some are optional):With just a few lines of code, you can start using hybrid search. You can run a test query in the Weaviate console using GraphQL. The query is, “Fisherman that catches salmon” (similar to the example above). When we set the alpha to 0.5 it is equally weighing the dense and sparse vector results.This article explained the concept of hybrid search. In the context of the Weaviate vector database, hybrid search is a combination of keyword search and semantic search. By combining these two types of searches, hybrid search queries can lead to more relevant results in use cases where you want to leverage the power of semantic search but also rely on exact matches of certain keywords, such as product or brand names. Check out the documentation for more information on hybrid search!Check out the Quickstart tutorial, and begin building amazing apps with the free trial of Weaviate Cloud (WCD).GitHubForumSlackX (Twitter)Sign up for our bi-weekly newsletter to stay updated!', metadata={'source': 'https://weaviate.io/blog/hybrid-search-explained'})]\n"
     ]
    }
   ],
   "source": [
    "# https://python.langchain.com/v0.2/docs/integrations/document_loaders/web_base/\n",
    "from langchain_community.document_loaders import WebBaseLoader\n",
    "import bs4\n",
    "\n",
    "## load, chunk, index contexts of page\n",
    "loader=WebBaseLoader(web_paths=[\n",
    "    \"https://lilianweng.github.io/posts/2023-06-23-agent/\"],\n",
    "    # bs_kwargs=dict(parse_only=bs4.SoupStrainer(\"p\")) # parse only paragraphs\n",
    "    bs_kwargs=dict(parse_only=bs4.SoupStrainer(\n",
    "        class_=(\"post-title\", \"post-content\", \"post-header\") # parse only content of blog posts\n",
    "    ))\n",
    ")\n",
    "\n",
    "loader=WebBaseLoader(web_paths=[\n",
    "    \"https://weaviate.io/blog/hybrid-search-explained\"],\n",
    "    bs_kwargs=dict(parse_only=bs4.SoupStrainer( (\"header\", \"p\") ) )\n",
    ")\n",
    "\n",
    "docs=loader.load()\n",
    "print(docs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Other loader types:\n",
    "https://python.langchain.com/v0.2/docs/integrations/document_loaders/\n",
    "\n",
    "Example: PubMed loader\n",
    "https://python.langchain.com/v0.2/docs/integrations/document_loaders/pubmed/\n",
    "\n",
    "potentially other useful loaders:\n",
    "- HuggingFaceDatasetLoader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Too Many Requests, waiting for 0.20 seconds...\n",
      "{'Copyright Information': 'Copyright © 2017 Elsevier Inc. All rights reserved.',\n",
      " 'Published': '--',\n",
      " 'Title': 'Chronic Cigarette Smoke-Induced Epigenomic Changes Precede '\n",
      "          'Sensitization of Bronchial Epithelial Cells to Single-Step '\n",
      "          'Transformation by KRAS Mutations.',\n",
      " 'uid': '28898697'}\n",
      "We define how chronic cigarette smoke-induced time-dependent epigenetic alterations can sensitize human bronchial epithelial cells for transformation by a single oncogene. The smoke-induced chromatin changes include initial repressive polycomb marking of genes, later manifesting abnormal DNA methylation by 10 months. At this time, cells exhibit epithelial-to-mesenchymal changes, anchorage-independent growth, and upregulated RAS/MAPK signaling with silencing of hypermethylated genes, which normally inhibit these pathways and are associated with smoking-related non-small cell lung cancer. These cells, in the absence of any driver gene mutations, now transform by introducing a single KRAS mutation and form adenosquamous lung carcinomas in mice. Thus, epigenetic abnormalities may prime for changing oncogene senescence to addiction for a single key oncogene involved in lung cancer initiation.\n"
     ]
    }
   ],
   "source": [
    "from langchain_community.document_loaders import PubMedLoader\n",
    "from pprint import pprint\n",
    "\n",
    "loader = PubMedLoader(query=\"Chronic Cigarette Smoke-Induced Epigenomic Changes\")\n",
    "docs = loader.load()\n",
    "\n",
    "len(docs)\n",
    "pprint(docs[0].metadata)\n",
    "print(docs[0].page_content)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Transformation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Convert PDF to chunks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Document(page_content='Chronic Cigarette Smoke-Induced Epigenomic Changes Precede \\nSensitization of Bronchial Epithelial Cells to Single Step \\nTransformation by KRAS  Mutations\\nMichelle Vaz1, Stephen Y Hwang1, Ioannis Kagiampakis1, Jillian Phallen1, Ashwini Patil2, \\nHeather M O’Hagan3, Lauren Murphy1, Cynthia A Zahnow1, Edward Gabrielson4, Victor E \\nVelculescu1, Hariharan P Easwaran1,5, and Stephen B Baylin1,5,6\\n1Department of Oncology, The Sidney Kimmel Comprehensive Cancer Center, The Johns \\nHopkins University School of Medicine, Baltimore, MD 21287, USA.\\n2Krieger School of Arts and Sciences, Baltimore, MD 21218, USA\\n3Medical Sciences, Indiana University School of Medicine, Bloomington, IN 47405, USA Melvin \\nand Bren Simon Cancer Center, Indianapolis, IN 46202, USA\\n4Department of Pathology, The Johns Hopkins University School of Medicine, Baltimore, MD \\n21287\\nSUMMARY\\nWe define how chronic cigarette smoke-induced time-dependent epigenetic alterations can', metadata={'source': 'cancercell.pdf', 'page': 0}),\n",
      " Document(page_content='SUMMARY\\nWe define how chronic cigarette smoke-induced time-dependent epigenetic alterations can \\nsensitize human bronchial epithelial cells for transformation by a single oncogene. The smoke-\\ninduced chromatin changes include initial repressive polycomb marking of genes, later \\nmanifesting abnormal DNA methylation by 10 months. At this time, cells exhibit epithelial to \\nmesenchymal changes, anchorage-independent growth and upregulated RAS/MAPK signaling \\nwith silencing of hyper-methylated genes, which normally inhibit these pathways and are \\nassociated with smoking related NSCLC. These cells, in the absence of any driver gene mutations, \\nnow transform by introducing a single KRAS  mutation and form adeno-squamous lung \\ncarcinomas in mice. Thus, epigenetic abnormalities may prime for changing oncogene senescence \\nto addiction for a single key oncogene involved in lung cancer initiation.\\neTOC blurb/In Brief', metadata={'source': 'cancercell.pdf', 'page': 0}),\n",
      " Document(page_content=\"to addiction for a single key oncogene involved in lung cancer initiation.\\neTOC blurb/In Brief\\n5To whom all correspondence should be addressed: Stephen B. Baylin , Department of Oncology, The Sidney Kimmel Comprehensive \\nCancer Center at Johns Hopkins, sbaylin@jhmi.edu, Hariharan P Easwaran , Department of Oncology, The Sidney Kimmel \\nComprehensive Cancer Center at Johns Hopkins, Baltimore, MD 21287, USA. heaswar2@jhmi.edu.5Lead Contact, Stephen B. Baylin , Department of Oncology, The Sidney Kimmel Comprehensive Cancer Center at Johns Hopkins, \\nBaltimore, MD 21287, USA. sbaylin@jhmi.edu.\\nPublisher's Disclaimer: This is a PDF file of an unedited manuscript that has been accepted for publication. As a service to our \\ncustomers we are providing this early version of the manuscript. The manuscript will undergo copyediting, typesetting, and review of \\nthe resulting proof before it is published in its final citable form. Please note that during the production process errors may be\", metadata={'source': 'cancercell.pdf', 'page': 0})]\n"
     ]
    }
   ],
   "source": [
    "from langchain.text_splitter import RecursiveCharacterTextSplitter\n",
    "\n",
    "text_splitter = RecursiveCharacterTextSplitter(\n",
    "    chunk_size=1000, # split text into chunks of 1000 characters\n",
    "    chunk_overlap=100 # overlap chunks by 100 characters\n",
    ")\n",
    "\n",
    "documents=text_splitter.split_documents(pdf_docs)\n",
    "pprint(documents[:3])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Create embeddings from chunks\n",
    "Store vectors in Chroma"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_community.embeddings import OpenAIEmbeddings\n",
    "from langchain_community.vectorstores import Chroma\n",
    "\n",
    "db_chroma = Chroma.from_documents(documents, embedding=OpenAIEmbeddings())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Store vectors in FAISS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_community.embeddings import OpenAIEmbeddings\n",
    "from langchain_community.vectorstores import FAISS\n",
    "\n",
    "db_faiss = FAISS.from_documents(documents, embedding=OpenAIEmbeddings())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Store vectors in Qdrant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [
    {
     "ename": "ResponseHandlingException",
     "evalue": "[Errno 61] Connection refused",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mConnectError\u001b[0m                              Traceback (most recent call last)",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/httpx/_transports/default.py:69\u001b[0m, in \u001b[0;36mmap_httpcore_exceptions\u001b[0;34m()\u001b[0m\n\u001b[1;32m     68\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m---> 69\u001b[0m     \u001b[38;5;28;01myield\u001b[39;00m\n\u001b[1;32m     70\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m exc:\n",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/httpx/_transports/default.py:233\u001b[0m, in \u001b[0;36mHTTPTransport.handle_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    232\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m map_httpcore_exceptions():\n\u001b[0;32m--> 233\u001b[0m     resp \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_pool\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhandle_request\u001b[49m\u001b[43m(\u001b[49m\u001b[43mreq\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    235\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(resp\u001b[38;5;241m.\u001b[39mstream, typing\u001b[38;5;241m.\u001b[39mIterable)\n",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/httpcore/_sync/connection_pool.py:216\u001b[0m, in \u001b[0;36mConnectionPool.handle_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    215\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_close_connections(closing)\n\u001b[0;32m--> 216\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m exc \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    218\u001b[0m \u001b[38;5;66;03m# Return the response. Note that in this case we still have to manage\u001b[39;00m\n\u001b[1;32m    219\u001b[0m \u001b[38;5;66;03m# the point at which the response is closed.\u001b[39;00m\n",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/httpcore/_sync/connection_pool.py:196\u001b[0m, in \u001b[0;36mConnectionPool.handle_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    194\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    195\u001b[0m     \u001b[38;5;66;03m# Send the request on the assigned connection.\u001b[39;00m\n\u001b[0;32m--> 196\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[43mconnection\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhandle_request\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    197\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpool_request\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\n\u001b[1;32m    198\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    199\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m ConnectionNotAvailable:\n\u001b[1;32m    200\u001b[0m     \u001b[38;5;66;03m# In some cases a connection may initially be available to\u001b[39;00m\n\u001b[1;32m    201\u001b[0m     \u001b[38;5;66;03m# handle a request, but then become unavailable.\u001b[39;00m\n\u001b[1;32m    202\u001b[0m     \u001b[38;5;66;03m#\u001b[39;00m\n\u001b[1;32m    203\u001b[0m     \u001b[38;5;66;03m# In this case we clear the connection and try again.\u001b[39;00m\n",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/httpcore/_sync/connection.py:99\u001b[0m, in \u001b[0;36mHTTPConnection.handle_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m     98\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_connect_failed \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[0;32m---> 99\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m exc\n\u001b[1;32m    101\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_connection\u001b[38;5;241m.\u001b[39mhandle_request(request)\n",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/httpcore/_sync/connection.py:76\u001b[0m, in \u001b[0;36mHTTPConnection.handle_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m     75\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_connection \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m---> 76\u001b[0m     stream \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_connect\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     78\u001b[0m     ssl_object \u001b[38;5;241m=\u001b[39m stream\u001b[38;5;241m.\u001b[39mget_extra_info(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mssl_object\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/httpcore/_sync/connection.py:122\u001b[0m, in \u001b[0;36mHTTPConnection._connect\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    121\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m Trace(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mconnect_tcp\u001b[39m\u001b[38;5;124m\"\u001b[39m, logger, request, kwargs) \u001b[38;5;28;01mas\u001b[39;00m trace:\n\u001b[0;32m--> 122\u001b[0m     stream \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_network_backend\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect_tcp\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    123\u001b[0m     trace\u001b[38;5;241m.\u001b[39mreturn_value \u001b[38;5;241m=\u001b[39m stream\n",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/httpcore/_backends/sync.py:205\u001b[0m, in \u001b[0;36mSyncBackend.connect_tcp\u001b[0;34m(self, host, port, timeout, local_address, socket_options)\u001b[0m\n\u001b[1;32m    200\u001b[0m exc_map: ExceptionMapping \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    201\u001b[0m     socket\u001b[38;5;241m.\u001b[39mtimeout: ConnectTimeout,\n\u001b[1;32m    202\u001b[0m     \u001b[38;5;167;01mOSError\u001b[39;00m: ConnectError,\n\u001b[1;32m    203\u001b[0m }\n\u001b[0;32m--> 205\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m map_exceptions(exc_map):\n\u001b[1;32m    206\u001b[0m     sock \u001b[38;5;241m=\u001b[39m socket\u001b[38;5;241m.\u001b[39mcreate_connection(\n\u001b[1;32m    207\u001b[0m         address,\n\u001b[1;32m    208\u001b[0m         timeout,\n\u001b[1;32m    209\u001b[0m         source_address\u001b[38;5;241m=\u001b[39msource_address,\n\u001b[1;32m    210\u001b[0m     )\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/contextlib.py:153\u001b[0m, in \u001b[0;36m_GeneratorContextManager.__exit__\u001b[0;34m(self, typ, value, traceback)\u001b[0m\n\u001b[1;32m    152\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 153\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgen\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mthrow\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtyp\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mvalue\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtraceback\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    154\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mStopIteration\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m exc:\n\u001b[1;32m    155\u001b[0m     \u001b[38;5;66;03m# Suppress StopIteration *unless* it's the same exception that\u001b[39;00m\n\u001b[1;32m    156\u001b[0m     \u001b[38;5;66;03m# was passed to throw().  This prevents a StopIteration\u001b[39;00m\n\u001b[1;32m    157\u001b[0m     \u001b[38;5;66;03m# raised inside the \"with\" statement from being suppressed.\u001b[39;00m\n",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/httpcore/_exceptions.py:14\u001b[0m, in \u001b[0;36mmap_exceptions\u001b[0;34m(map)\u001b[0m\n\u001b[1;32m     13\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(exc, from_exc):\n\u001b[0;32m---> 14\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m to_exc(exc) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mexc\u001b[39;00m\n\u001b[1;32m     15\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m\n",
      "\u001b[0;31mConnectError\u001b[0m: [Errno 61] Connection refused",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mConnectError\u001b[0m                              Traceback (most recent call last)",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/qdrant_client/http/api_client.py:103\u001b[0m, in \u001b[0;36mApiClient.send_inner\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    102\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 103\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_client\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    104\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/httpx/_client.py:914\u001b[0m, in \u001b[0;36mClient.send\u001b[0;34m(self, request, stream, auth, follow_redirects)\u001b[0m\n\u001b[1;32m    912\u001b[0m auth \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_build_request_auth(request, auth)\n\u001b[0;32m--> 914\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_send_handling_auth\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    915\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    916\u001b[0m \u001b[43m    \u001b[49m\u001b[43mauth\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mauth\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    917\u001b[0m \u001b[43m    \u001b[49m\u001b[43mfollow_redirects\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfollow_redirects\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    918\u001b[0m \u001b[43m    \u001b[49m\u001b[43mhistory\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m[\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    919\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    920\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/httpx/_client.py:942\u001b[0m, in \u001b[0;36mClient._send_handling_auth\u001b[0;34m(self, request, auth, follow_redirects, history)\u001b[0m\n\u001b[1;32m    941\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[0;32m--> 942\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_send_handling_redirects\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    943\u001b[0m \u001b[43m        \u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    944\u001b[0m \u001b[43m        \u001b[49m\u001b[43mfollow_redirects\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfollow_redirects\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    945\u001b[0m \u001b[43m        \u001b[49m\u001b[43mhistory\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mhistory\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    946\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    947\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/httpx/_client.py:979\u001b[0m, in \u001b[0;36mClient._send_handling_redirects\u001b[0;34m(self, request, follow_redirects, history)\u001b[0m\n\u001b[1;32m    977\u001b[0m     hook(request)\n\u001b[0;32m--> 979\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_send_single_request\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    980\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/httpx/_client.py:1015\u001b[0m, in \u001b[0;36mClient._send_single_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m   1014\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m request_context(request\u001b[38;5;241m=\u001b[39mrequest):\n\u001b[0;32m-> 1015\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[43mtransport\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhandle_request\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1017\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(response\u001b[38;5;241m.\u001b[39mstream, SyncByteStream)\n",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/httpx/_transports/default.py:232\u001b[0m, in \u001b[0;36mHTTPTransport.handle_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    220\u001b[0m req \u001b[38;5;241m=\u001b[39m httpcore\u001b[38;5;241m.\u001b[39mRequest(\n\u001b[1;32m    221\u001b[0m     method\u001b[38;5;241m=\u001b[39mrequest\u001b[38;5;241m.\u001b[39mmethod,\n\u001b[1;32m    222\u001b[0m     url\u001b[38;5;241m=\u001b[39mhttpcore\u001b[38;5;241m.\u001b[39mURL(\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    230\u001b[0m     extensions\u001b[38;5;241m=\u001b[39mrequest\u001b[38;5;241m.\u001b[39mextensions,\n\u001b[1;32m    231\u001b[0m )\n\u001b[0;32m--> 232\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m map_httpcore_exceptions():\n\u001b[1;32m    233\u001b[0m     resp \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_pool\u001b[38;5;241m.\u001b[39mhandle_request(req)\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/contextlib.py:153\u001b[0m, in \u001b[0;36m_GeneratorContextManager.__exit__\u001b[0;34m(self, typ, value, traceback)\u001b[0m\n\u001b[1;32m    152\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 153\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgen\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mthrow\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtyp\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mvalue\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtraceback\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    154\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mStopIteration\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m exc:\n\u001b[1;32m    155\u001b[0m     \u001b[38;5;66;03m# Suppress StopIteration *unless* it's the same exception that\u001b[39;00m\n\u001b[1;32m    156\u001b[0m     \u001b[38;5;66;03m# was passed to throw().  This prevents a StopIteration\u001b[39;00m\n\u001b[1;32m    157\u001b[0m     \u001b[38;5;66;03m# raised inside the \"with\" statement from being suppressed.\u001b[39;00m\n",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/httpx/_transports/default.py:86\u001b[0m, in \u001b[0;36mmap_httpcore_exceptions\u001b[0;34m()\u001b[0m\n\u001b[1;32m     85\u001b[0m message \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mstr\u001b[39m(exc)\n\u001b[0;32m---> 86\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m mapped_exc(message) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mexc\u001b[39;00m\n",
      "\u001b[0;31mConnectError\u001b[0m: [Errno 61] Connection refused",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mResponseHandlingException\u001b[0m                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[183], line 6\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain_community\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mvectorstores\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m Qdrant\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mqdrant_client\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m QdrantClient\n\u001b[0;32m----> 6\u001b[0m db_qdrant \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mawait\u001b[39;00m Qdrant\u001b[38;5;241m.\u001b[39mafrom_documents(documents, embedding\u001b[38;5;241m=\u001b[39mOpenAIEmbeddings())\n",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/langchain_core/vectorstores.py:652\u001b[0m, in \u001b[0;36mVectorStore.afrom_documents\u001b[0;34m(cls, documents, embedding, **kwargs)\u001b[0m\n\u001b[1;32m    650\u001b[0m texts \u001b[38;5;241m=\u001b[39m [d\u001b[38;5;241m.\u001b[39mpage_content \u001b[38;5;28;01mfor\u001b[39;00m d \u001b[38;5;129;01min\u001b[39;00m documents]\n\u001b[1;32m    651\u001b[0m metadatas \u001b[38;5;241m=\u001b[39m [d\u001b[38;5;241m.\u001b[39mmetadata \u001b[38;5;28;01mfor\u001b[39;00m d \u001b[38;5;129;01min\u001b[39;00m documents]\n\u001b[0;32m--> 652\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mawait\u001b[39;00m \u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39mafrom_texts(texts, embedding, metadatas\u001b[38;5;241m=\u001b[39mmetadatas, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/langchain_community/vectorstores/qdrant.py:56\u001b[0m, in \u001b[0;36msync_call_fallback.<locals>.wrapper\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m     53\u001b[0m \u001b[38;5;129m@functools\u001b[39m\u001b[38;5;241m.\u001b[39mwraps(method)\n\u001b[1;32m     54\u001b[0m \u001b[38;5;28;01masync\u001b[39;00m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mwrapper\u001b[39m(\u001b[38;5;28mself\u001b[39m: Any, \u001b[38;5;241m*\u001b[39margs: Any, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: Any) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m Any:\n\u001b[1;32m     55\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m---> 56\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mawait\u001b[39;00m method(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m     57\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mNotImplementedError\u001b[39;00m:\n\u001b[1;32m     58\u001b[0m         \u001b[38;5;66;03m# If the async method is not implemented, call the synchronous method\u001b[39;00m\n\u001b[1;32m     59\u001b[0m         \u001b[38;5;66;03m# by removing the first letter from the method name. For example,\u001b[39;00m\n\u001b[1;32m     60\u001b[0m         \u001b[38;5;66;03m# if the async method is called ``aaad_texts``, the synchronous method\u001b[39;00m\n\u001b[1;32m     61\u001b[0m         \u001b[38;5;66;03m# will be called ``aad_texts``.\u001b[39;00m\n\u001b[1;32m     62\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mawait\u001b[39;00m run_in_executor(\n\u001b[1;32m     63\u001b[0m             \u001b[38;5;28;01mNone\u001b[39;00m, \u001b[38;5;28mgetattr\u001b[39m(\u001b[38;5;28mself\u001b[39m, method\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m[\u001b[38;5;241m1\u001b[39m:]), \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs\n\u001b[1;32m     64\u001b[0m         )\n",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/langchain_community/vectorstores/qdrant.py:1560\u001b[0m, in \u001b[0;36mQdrant.afrom_texts\u001b[0;34m(cls, texts, embedding, metadatas, ids, location, url, port, grpc_port, prefer_grpc, https, api_key, prefix, timeout, host, path, collection_name, distance_func, content_payload_key, metadata_payload_key, vector_name, batch_size, shard_number, replication_factor, write_consistency_factor, on_disk_payload, hnsw_config, optimizers_config, wal_config, quantization_config, init_from, on_disk, force_recreate, **kwargs)\u001b[0m\n\u001b[1;32m   1419\u001b[0m \u001b[38;5;129m@classmethod\u001b[39m\n\u001b[1;32m   1420\u001b[0m \u001b[38;5;129m@sync_call_fallback\u001b[39m\n\u001b[1;32m   1421\u001b[0m \u001b[38;5;28;01masync\u001b[39;00m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mafrom_texts\u001b[39m(\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1455\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: Any,\n\u001b[1;32m   1456\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m Qdrant:\n\u001b[1;32m   1457\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Construct Qdrant wrapper from a list of texts.\u001b[39;00m\n\u001b[1;32m   1458\u001b[0m \n\u001b[1;32m   1459\u001b[0m \u001b[38;5;124;03m    Args:\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1558\u001b[0m \u001b[38;5;124;03m            qdrant = await Qdrant.afrom_texts(texts, embeddings, \"localhost\")\u001b[39;00m\n\u001b[1;32m   1559\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 1560\u001b[0m     qdrant \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mawait\u001b[39;00m \u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39maconstruct_instance(\n\u001b[1;32m   1561\u001b[0m         texts,\n\u001b[1;32m   1562\u001b[0m         embedding,\n\u001b[1;32m   1563\u001b[0m         location,\n\u001b[1;32m   1564\u001b[0m         url,\n\u001b[1;32m   1565\u001b[0m         port,\n\u001b[1;32m   1566\u001b[0m         grpc_port,\n\u001b[1;32m   1567\u001b[0m         prefer_grpc,\n\u001b[1;32m   1568\u001b[0m         https,\n\u001b[1;32m   1569\u001b[0m         api_key,\n\u001b[1;32m   1570\u001b[0m         prefix,\n\u001b[1;32m   1571\u001b[0m         timeout,\n\u001b[1;32m   1572\u001b[0m         host,\n\u001b[1;32m   1573\u001b[0m         path,\n\u001b[1;32m   1574\u001b[0m         collection_name,\n\u001b[1;32m   1575\u001b[0m         distance_func,\n\u001b[1;32m   1576\u001b[0m         content_payload_key,\n\u001b[1;32m   1577\u001b[0m         metadata_payload_key,\n\u001b[1;32m   1578\u001b[0m         vector_name,\n\u001b[1;32m   1579\u001b[0m         shard_number,\n\u001b[1;32m   1580\u001b[0m         replication_factor,\n\u001b[1;32m   1581\u001b[0m         write_consistency_factor,\n\u001b[1;32m   1582\u001b[0m         on_disk_payload,\n\u001b[1;32m   1583\u001b[0m         hnsw_config,\n\u001b[1;32m   1584\u001b[0m         optimizers_config,\n\u001b[1;32m   1585\u001b[0m         wal_config,\n\u001b[1;32m   1586\u001b[0m         quantization_config,\n\u001b[1;32m   1587\u001b[0m         init_from,\n\u001b[1;32m   1588\u001b[0m         on_disk,\n\u001b[1;32m   1589\u001b[0m         force_recreate,\n\u001b[1;32m   1590\u001b[0m         \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m   1591\u001b[0m     )\n\u001b[1;32m   1592\u001b[0m     \u001b[38;5;28;01mawait\u001b[39;00m qdrant\u001b[38;5;241m.\u001b[39maadd_texts(texts, metadatas, ids, batch_size)\n\u001b[1;32m   1593\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m qdrant\n",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/langchain_community/vectorstores/qdrant.py:1832\u001b[0m, in \u001b[0;36mQdrant.aconstruct_instance\u001b[0;34m(cls, texts, embedding, location, url, port, grpc_port, prefer_grpc, https, api_key, prefix, timeout, host, path, collection_name, distance_func, content_payload_key, metadata_payload_key, vector_name, shard_number, replication_factor, write_consistency_factor, on_disk_payload, hnsw_config, optimizers_config, wal_config, quantization_config, init_from, on_disk, force_recreate, **kwargs)\u001b[0m\n\u001b[1;32m   1827\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m\n\u001b[1;32m   1829\u001b[0m \u001b[38;5;66;03m# Get the vector configuration of the existing collection and vector, if it\u001b[39;00m\n\u001b[1;32m   1830\u001b[0m \u001b[38;5;66;03m# was specified. If the old configuration does not match the current one,\u001b[39;00m\n\u001b[1;32m   1831\u001b[0m \u001b[38;5;66;03m# an exception is being thrown.\u001b[39;00m\n\u001b[0;32m-> 1832\u001b[0m collection_info \u001b[38;5;241m=\u001b[39m \u001b[43mclient\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_collection\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcollection_name\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcollection_name\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1833\u001b[0m current_vector_config \u001b[38;5;241m=\u001b[39m collection_info\u001b[38;5;241m.\u001b[39mconfig\u001b[38;5;241m.\u001b[39mparams\u001b[38;5;241m.\u001b[39mvectors\n\u001b[1;32m   1834\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(current_vector_config, \u001b[38;5;28mdict\u001b[39m) \u001b[38;5;129;01mand\u001b[39;00m vector_name \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/qdrant_client/qdrant_client.py:1578\u001b[0m, in \u001b[0;36mQdrantClient.get_collection\u001b[0;34m(self, collection_name, **kwargs)\u001b[0m\n\u001b[1;32m   1568\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Get detailed information about specified existing collection\u001b[39;00m\n\u001b[1;32m   1569\u001b[0m \n\u001b[1;32m   1570\u001b[0m \u001b[38;5;124;03mArgs:\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1574\u001b[0m \u001b[38;5;124;03m    Detailed information about the collection\u001b[39;00m\n\u001b[1;32m   1575\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   1576\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(kwargs) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m, \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mUnknown arguments: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mlist\u001b[39m(kwargs\u001b[38;5;241m.\u001b[39mkeys())\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m-> 1578\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_client\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_collection\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcollection_name\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcollection_name\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/qdrant_client/qdrant_remote.py:2050\u001b[0m, in \u001b[0;36mQdrantRemote.get_collection\u001b[0;34m(self, collection_name, **kwargs)\u001b[0m\n\u001b[1;32m   2043\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_prefer_grpc:\n\u001b[1;32m   2044\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m GrpcToRest\u001b[38;5;241m.\u001b[39mconvert_collection_info(\n\u001b[1;32m   2045\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mgrpc_collections\u001b[38;5;241m.\u001b[39mGet(\n\u001b[1;32m   2046\u001b[0m             grpc\u001b[38;5;241m.\u001b[39mGetCollectionInfoRequest(collection_name\u001b[38;5;241m=\u001b[39mcollection_name),\n\u001b[1;32m   2047\u001b[0m             timeout\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_timeout,\n\u001b[1;32m   2048\u001b[0m         )\u001b[38;5;241m.\u001b[39mresult\n\u001b[1;32m   2049\u001b[0m     )\n\u001b[0;32m-> 2050\u001b[0m result: Optional[types\u001b[38;5;241m.\u001b[39mCollectionInfo] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhttp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcollections_api\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_collection\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   2051\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcollection_name\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcollection_name\u001b[49m\n\u001b[1;32m   2052\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mresult\n\u001b[1;32m   2053\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m result \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mGet collection returned None\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   2054\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/qdrant_client/http/api/collections_api.py:1314\u001b[0m, in \u001b[0;36mSyncCollectionsApi.get_collection\u001b[0;34m(self, collection_name)\u001b[0m\n\u001b[1;32m   1307\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_collection\u001b[39m(\n\u001b[1;32m   1308\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m   1309\u001b[0m     collection_name: \u001b[38;5;28mstr\u001b[39m,\n\u001b[1;32m   1310\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m m\u001b[38;5;241m.\u001b[39mInlineResponse2006:\n\u001b[1;32m   1311\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   1312\u001b[0m \u001b[38;5;124;03m    Get detailed information about specified existing collection\u001b[39;00m\n\u001b[1;32m   1313\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 1314\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_build_for_get_collection\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1315\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcollection_name\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcollection_name\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1316\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/qdrant_client/http/api/collections_api.py:397\u001b[0m, in \u001b[0;36m_CollectionsApi._build_for_get_collection\u001b[0;34m(self, collection_name)\u001b[0m\n\u001b[1;32m    392\u001b[0m path_params \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    393\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcollection_name\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;28mstr\u001b[39m(collection_name),\n\u001b[1;32m    394\u001b[0m }\n\u001b[1;32m    396\u001b[0m headers \u001b[38;5;241m=\u001b[39m {}\n\u001b[0;32m--> 397\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapi_client\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    398\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtype_\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mm\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mInlineResponse2006\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    399\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mGET\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    400\u001b[0m \u001b[43m    \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m/collections/\u001b[39;49m\u001b[38;5;132;43;01m{collection_name}\u001b[39;49;00m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    401\u001b[0m \u001b[43m    \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mheaders\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01melse\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    402\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpath_params\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpath_params\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    403\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/qdrant_client/http/api_client.py:76\u001b[0m, in \u001b[0;36mApiClient.request\u001b[0;34m(self, type_, method, url, path_params, **kwargs)\u001b[0m\n\u001b[1;32m     74\u001b[0m     kwargs[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtimeout\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mint\u001b[39m(kwargs[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mparams\u001b[39m\u001b[38;5;124m\"\u001b[39m][\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtimeout\u001b[39m\u001b[38;5;124m\"\u001b[39m])\n\u001b[1;32m     75\u001b[0m request \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_client\u001b[38;5;241m.\u001b[39mbuild_request(method, url, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[0;32m---> 76\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtype_\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/qdrant_client/http/api_client.py:93\u001b[0m, in \u001b[0;36mApiClient.send\u001b[0;34m(self, request, type_)\u001b[0m\n\u001b[1;32m     92\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21msend\u001b[39m(\u001b[38;5;28mself\u001b[39m, request: Request, type_: Type[T]) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m T:\n\u001b[0;32m---> 93\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmiddleware\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend_inner\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     94\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m response\u001b[38;5;241m.\u001b[39mstatus_code \u001b[38;5;129;01min\u001b[39;00m [\u001b[38;5;241m200\u001b[39m, \u001b[38;5;241m201\u001b[39m, \u001b[38;5;241m202\u001b[39m]:\n\u001b[1;32m     95\u001b[0m         \u001b[38;5;28;01mtry\u001b[39;00m:\n",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/qdrant_client/http/api_client.py:202\u001b[0m, in \u001b[0;36mBaseMiddleware.__call__\u001b[0;34m(self, request, call_next)\u001b[0m\n\u001b[1;32m    201\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__call__\u001b[39m(\u001b[38;5;28mself\u001b[39m, request: Request, call_next: Send) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m Response:\n\u001b[0;32m--> 202\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mcall_next\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Desktop/github/gnawhnehpets/langchain/venv/lib/python3.10/site-packages/qdrant_client/http/api_client.py:105\u001b[0m, in \u001b[0;36mApiClient.send_inner\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    103\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_client\u001b[38;5;241m.\u001b[39msend(request)\n\u001b[1;32m    104\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[0;32m--> 105\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m ResponseHandlingException(e)\n\u001b[1;32m    106\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m response\n",
      "\u001b[0;31mResponseHandlingException\u001b[0m: [Errno 61] Connection refused"
     ]
    }
   ],
   "source": [
    "from langchain_community.embeddings import OpenAIEmbeddings\n",
    "from langchain_community.vectorstores import Qdrant\n",
    "from qdrant_client import QdrantClient\n",
    "\n",
    "\n",
    "# db_qdrant = await Qdrant.afrom_documents(documents, embedding=OpenAIEmbeddings())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Query vector store"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "response1: Figure 6. Phenotypes of CSC exposed versus control cells before and after expression of mutant \n",
      "KRASV12\n",
      "(A) Soft agar assays to assess the anchorage-independent growth ability of HBEC cells \n",
      "following 10 months of CSC exposure. (B) Quantitation of soft agar colonies in panel A. \n",
      "Values indicate average number of colonies ± SD from replicate 1 and 2 at the indicated time \n",
      "points. (C) Phase contrast photomicrographs showing morphological changes in control and \n",
      "CSC treated cells when switched from defined serum-free to serum-containing medium for \n",
      "96 hr. Scale bar = 50 µm. (D) Immunofluorescent staining of E-cadherin and Vimentin in \n",
      "cells grown in serum-containing medium for 96 hr. Scale bar = 100 µm. (E) Immunoblot for \n",
      "EMT markers in cells (CSC and Con) and soft agar clones obtained from 15 month CSC Vaz et al. Page 34\n",
      "Cancer Cell . Author manuscript; available in PMC 2018 September 11.\n",
      "Author Manuscript Author Manuscript Author Manuscript Author Manuscript\n",
      "response2: and this is accompanied by a stark increase in vimentin, a key protein upregulated in EMT, \n",
      "and a decrease in E-cadherin (Figure 6D and 6E). However, despite these changes, neither \n",
      "the CSC-exposed cells nor the isolated soft agar clones could form tumors in vivo.\n",
      "We hypothesized that the above progressive development of the epigenetic, signaling \n",
      "pathway and morphogenic changes induced by CSC exposure might sensitize the cells to \n",
      "transformation by a single oncogenic event. This would be highly significant since earlier \n",
      "studies have shown that complete transformation of HBEC cells requires introduction of at \n",
      "least three genetic events: shRNA-mediated downregulation of TP53 paired with \n",
      "introduction of both KRAS  mutations and CMYC overexpression ( Sato et al., 2013 ). We \n",
      "introduced individually either KRASV12 (Figure 6F) or downregulate the tumor suppressor \n",
      "TP53 (Figure 6G) in 6 and 15 month CSC treated cells. Only KRASV12 alters the 15 month Vaz et al. Page 7\n"
     ]
    }
   ],
   "source": [
    "query=\"When do the cells become anchorage independent and form clones in soft agar?\"\n",
    "response1 = db_chroma.similarity_search(query)\n",
    "print(f'response1: {response1[1].page_content}')\n",
    "\n",
    "response2 = db_faiss.similarity_search(query)\n",
    "print(f'response2: {response2[1].page_content}')\n",
    "\n",
    "# response3 = await db_qdrant.asimilarity_search(query)\n",
    "# print(docs[0].page_content)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Querying"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ollama(model='llama3')"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from langchain_community.llms import Ollama\n",
    "\n",
    "llm=Ollama(model=\"llama3\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Design prompt template"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_core.prompts import ChatPromptTemplate\n",
    "\n",
    "\n",
    "prompt_template=ChatPromptTemplate.from_template(\"\"\"\n",
    "Answer the following quetsion based on the only information provided in the text.\n",
    "Think step-by-step before providing a detailed answer in 100 words or less.\n",
    "The question should be easy enough for a child to understand.\n",
    "<context>{context}</context>\n",
    "Question: {input}\n",
    "\"\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Creating document chain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain.chains.combine_documents import create_stuff_documents_chain\n",
    "\n",
    "document_chain = create_stuff_documents_chain(llm, prompt_template)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Adding retriever (interface to VS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "retriever = db_faiss.as_retriever()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Creating retriever chain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain.chains import create_retrieval_chain\n",
    "\n",
    "retrieval_chain = create_retrieval_chain(retriever, document_chain)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Invoke the chain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "retrieval_chain.invoke({\"input\": \"When do the cells become anchorage independent and form clones in soft agar?\"})"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
